---
title: homework 3 - population structure and GWAS
author: Haky Im
date: 2025-04-11
categories:
  - bios25328
  - homework
format:
  html:
    toc: true
    code-fold: false
    code-summary: "Show the code"
    code-tools: true
    code-overflow: wrap
date-modified: last-modified
draft: false
---


## A. Complete the following assessments
- [Lecture 5 assessment quiz](https://forms.gle/ShSWpUvbT8MZ1bfDA) (15 points)


## B. Reproduce figures and analysis in lectures 5 and 6 (85 points)

1. Download HapMap3 chromosome 22 data (5pt)

  - You can find the full HapMap 3 genotype data with other useful information [here](https://uchicago.app.box.com/folder/155475906095?s=84de60jj4j9nyhl7nc7lylk4fcq12eka)

2. Download Growth Data (5pt)
    [here](https://github.com/hakyimlab/igrowth/blob/master/rawgrowth.txt)
  - if you want background information on this trait you can read [this paper](https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1002525))

3. Calculate PCs using plink (10pt)

4. Plot the first two principal components of the genotype matrix, color by population (10pt)

5. Run GWAS of growth with and without adjusting for PCs (20pt)
    1. show qqplot and manhattan plot for each (10pt)

6. Create HapMap Trios Relatedness Matrix (15pt)

7. generate a tileplot to visualize the population and relatedness patterns [here](https://docs.google.com/presentation/d/1I0-4u_6PAG5Q9LNi3mfQNnvZsVgf16tO/edit#slide=id.p13) (10pt)
    

**Links**

- You can borrow code from the lecture and the links below
    - [https://hakyimlab.github.io/hgen471/L6-population-structure.html](https://hakyimlab.github.io/hgen471/L6-population-structure.html)
    - [https://bios25328.hakyimlab.org/post/2020/02/06/l8-grm/](https://bios25328.hakyimlab.org/post/2020/02/06/l8-grm/)
- slides

#### The following code will directly download the bed/bim/fam files 

```bash
wget -O hapmap3_ch22.bed https://uchicago.box.com/shared/static/uu5fbx135qgp37spnrrsxrk4acuibkwh.bed
wget -O hapmap3_ch22.bim https://uchicago.box.com/shared/static/eh0thcbeo5sb1ywufrowtm3uqqgad0b2.bim
wget -O hapmap3_ch22.fam https://uchicago.box.com/shared/static/7py9smo190ucl6jrdwe0de8wjdc6rrwq.fam
```

#### How to download plink

```bash
wget https://s3.amazonaws.com/plink1-assets/plink_linux_x86_64_20220305.zip
##Use this link if you are on Rstudio Cloud otherwise go to the website
##https://zzz.bwh.harvard.edu/plink/download.shtml 
unzip plink_linux_x86_64_20220305.zip
mikdir bin 
mv plink bin
```

## C. Finish running the Marees GWAS tutorial 

- [ ] Finish [lab 3](/post/bios-25328/2024-04-05-lab-03/) and answer the following questions.


Using the output from the tutorial or using the commands you learned from it, answer the following questions. Show the command you used to create the result.

1. How many individuals are in the genotype file you downloaded? (5 pts)
2. Explain the contents of `.fam`, `.bim`, `.bed` files (5 pts)
3. Write the captions for the figures generated by the commands in 1_Main_script_QC_GWAS.txt and 3_Main_script_association_GWAS (5 pts per figure caption)
4. Explain what you accomplished with the tutorial and explain the results/figures you obtained. (20 pts)

