<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Haky Im">
<meta name="dcterms.date" content="2025-03-31">
<meta name="description" content="Lecture 4">

<title>lecture 4 - multiple testing correction - with code – HKI Courses</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../../">
<script src="../../../site_libs/quarto-html/quarto.js"></script>
<script src="../../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../../site_libs/quarto-html/quarto-syntax-highlighting-2f5df379a58b258e96c21c0638c20c03.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../../site_libs/bootstrap/bootstrap-dcc7d9e706e4adfbab8e3d6c7c60bab2.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="../../../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../../index.html">
    <span class="navbar-title">HKI Courses</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../index-qgt.html"> 
<span class="menu-text">QGT</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../index-bios-25328.html"> 
<span class="menu-text">BIOS25328</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../../index-logistics.html"> 
<span class="menu-text">Logistics</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#learning-objectives" id="toc-learning-objectives" class="nav-link active" data-scroll-target="#learning-objectives">Learning Objectives</a></li>
  <li><a href="#why-do-we-need-multiple-testing-correction" id="toc-why-do-we-need-multiple-testing-correction" class="nav-link" data-scroll-target="#why-do-we-need-multiple-testing-correction">Why Do We Need Multiple Testing Correction?</a></li>
  <li><a href="#what-do-p-values-look-like-under-null-and-alternative-hypotheses" id="toc-what-do-p-values-look-like-under-null-and-alternative-hypotheses" class="nav-link" data-scroll-target="#what-do-p-values-look-like-under-null-and-alternative-hypotheses">What Do P-values Look Like Under Null and Alternative Hypotheses?</a>
  <ul class="collapse">
  <li><a href="#simulate-data-null-vs.-alternative" id="toc-simulate-data-null-vs.-alternative" class="nav-link" data-scroll-target="#simulate-data-null-vs.-alternative">Simulate Data: Null vs.&nbsp;Alternative</a></li>
  <li><a href="#simulation-parameters" id="toc-simulation-parameters" class="nav-link" data-scroll-target="#simulation-parameters">Simulation Parameters</a></li>
  <li><a href="#generate-single-instance-of-data" id="toc-generate-single-instance-of-data" class="nav-link" data-scroll-target="#generate-single-instance-of-data">Generate Single Instance of Data</a></li>
  <li><a href="#visualize-the-data" id="toc-visualize-the-data" class="nav-link" data-scroll-target="#visualize-the-data">Visualize the Data</a></li>
  <li><a href="#test-associations-single-instance" id="toc-test-associations-single-instance" class="nav-link" data-scroll-target="#test-associations-single-instance">Test Associations (Single Instance)</a></li>
  <li><a href="#calculate-the-empirical-distribution-of-p-values" id="toc-calculate-the-empirical-distribution-of-p-values" class="nav-link" data-scroll-target="#calculate-the-empirical-distribution-of-p-values">Calculate the Empirical Distribution of P-values</a>
  <ul class="collapse">
  <li><a href="#convenience-function-fastlm" id="toc-convenience-function-fastlm" class="nav-link" data-scroll-target="#convenience-function-fastlm">Convenience Function fastlm</a></li>
  <li><a href="#run-10000-simulations" id="toc-run-10000-simulations" class="nav-link" data-scroll-target="#run-10000-simulations">Run 10,000 Simulations</a></li>
  <li><a href="#calculate-p-values-under-the-null" id="toc-calculate-p-values-under-the-null" class="nav-link" data-scroll-target="#calculate-p-values-under-the-null">Calculate P-values Under the Null</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#bonferroni-correction" id="toc-bonferroni-correction" class="nav-link" data-scroll-target="#bonferroni-correction">Bonferroni Correction</a></li>
  <li><a href="#mix-of-null-and-alternative-hypotheses" id="toc-mix-of-null-and-alternative-hypotheses" class="nav-link" data-scroll-target="#mix-of-null-and-alternative-hypotheses">Mix of Null and Alternative Hypotheses</a>
  <ul class="collapse">
  <li><a href="#create-mixed-data" id="toc-create-mixed-data" class="nav-link" data-scroll-target="#create-mixed-data">Create Mixed Data</a></li>
  <li><a href="#calculate-p-values-for-mixed-data" id="toc-calculate-p-values-for-mixed-data" class="nav-link" data-scroll-target="#calculate-p-values-for-mixed-data">Calculate P-values for Mixed Data</a></li>
  <li><a href="#interpreting-mixed-p-values" id="toc-interpreting-mixed-p-values" class="nav-link" data-scroll-target="#interpreting-mixed-p-values">Interpreting Mixed P-values</a></li>
  <li><a href="#false-discovery-rate-fdr-estimate-using-known-truth" id="toc-false-discovery-rate-fdr-estimate-using-known-truth" class="nav-link" data-scroll-target="#false-discovery-rate-fdr-estimate-using-known-truth">False Discovery Rate (FDR) Estimate (using known truth)</a></li>
  </ul></li>
  <li><a href="#common-approaches-to-control-type-i-errors" id="toc-common-approaches-to-control-type-i-errors" class="nav-link" data-scroll-target="#common-approaches-to-control-type-i-errors">Common Approaches to Control Type I Errors</a>
  <ul class="collapse">
  <li><a href="#table-of-simulation-results" id="toc-table-of-simulation-results" class="nav-link" data-scroll-target="#table-of-simulation-results">Table of Simulation Results</a></li>
  </ul></li>
  <li><a href="#use-qvalue-package-to-estimate-fdr" id="toc-use-qvalue-package-to-estimate-fdr" class="nav-link" data-scroll-target="#use-qvalue-package-to-estimate-fdr">Use qvalue Package to Estimate FDR</a>
  <ul class="collapse">
  <li><a href="#visualize-q-values" id="toc-visualize-q-values" class="nav-link" data-scroll-target="#visualize-q-values">Visualize Q-values</a></li>
  <li><a href="#how-do-q-values-and-p-values-relate" id="toc-how-do-q-values-and-p-values-relate" class="nav-link" data-scroll-target="#how-do-q-values-and-p-values-relate">How Do Q-values and P-values Relate?</a></li>
  <li><a href="#estimating-π₀-proportion-of-true-nulls" id="toc-estimating-π₀-proportion-of-true-nulls" class="nav-link" data-scroll-target="#estimating-π₀-proportion-of-true-nulls">Estimating π₀ (Proportion of True Nulls)</a></li>
  </ul></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">lecture 4 - multiple testing correction - with code</h1><button type="button" class="btn code-tools-button dropdown-toggle" id="quarto-code-tools-menu" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi"></i> Code</button><ul class="dropdown-menu dropdown-menu-end" aria-labelelledby="quarto-code-tools-menu"><li><a id="quarto-show-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Show All Code</a></li><li><a id="quarto-hide-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Hide All Code</a></li><li><hr class="dropdown-divider"></li><li><a id="quarto-view-source" class="dropdown-item" href="javascript:void(0)" role="button">View Source</a></li></ul></div></div>
  <div class="quarto-categories">
    <div class="quarto-category">bios25328</div>
    <div class="quarto-category">notebook</div>
  </div>
  </div>

<div>
  <div class="description">
    Lecture 4
  </div>
</div>


<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Haky Im </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">March 31, 2025</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<p>Find the lecture notes <a href="https://www.icloud.com/keynote/07fVVMHEvj8f2aG5zyyFfXisA#L4-Multiple-Testing-Correction-2025">here</a>.</p>
<section id="learning-objectives" class="level1">
<h1>Learning Objectives</h1>
<ul>
<li>Build intuition about p-values when multiple testing is performed via simulations.</li>
<li>Recognize the need for multiple testing correction.</li>
<li>Present methods to correct for multiple testing:
<ul>
<li>Bonferroni correction</li>
<li>FDR (false discovery rate)</li>
</ul></li>
</ul>
</section>
<section id="why-do-we-need-multiple-testing-correction" class="level1">
<h1>Why Do We Need Multiple Testing Correction?</h1>
<p>When performing many statistical tests simultaneously, the probability of finding a “significant” result just by chance (a Type I error or false positive) increases substantially. This is famously illustrated by the XKCD comic on significance (link to XKCD #882). We need methods to control the overall error rate across all tests.</p>
</section>
<section id="what-do-p-values-look-like-under-null-and-alternative-hypotheses" class="level1">
<h1>What Do P-values Look Like Under Null and Alternative Hypotheses?</h1>
<section id="simulate-data-null-vs.-alternative" class="level2">
<h2 class="anchored" data-anchor-id="simulate-data-null-vs.-alternative">Simulate Data: Null vs.&nbsp;Alternative</h2>
<p>We’ll start by simulating data under two scenarios:</p>
<ul>
<li>Null Hypothesis (Ynull): The outcome variable is independent of the predictor X.</li>
<li>Alternative Hypothesis (Yalt): The outcome variable depends on the predictor X via the relationship Yalt = X * beta + epsilon.</li>
</ul>
</section>
<section id="simulation-parameters" class="level2">
<h2 class="anchored" data-anchor-id="simulation-parameters">Simulation Parameters</h2>
<p>First, define some parameters for the simulation.</p>
<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Parameters</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>nsamp <span class="ot">=</span> <span class="dv">100</span>     <span class="co"># Number of samples</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>beta <span class="ot">=</span> <span class="dv">2</span>        <span class="co"># Effect size for the alternative hypothesis</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>h2 <span class="ot">=</span> <span class="fl">0.1</span>        <span class="co"># Proportion of variance in Yalt explained by X (related to signal strength)</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>sig2X <span class="ot">=</span> h2      <span class="co"># Variance of X (scaled for simplicity here)</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>sig2epsi <span class="ot">=</span> (<span class="dv">1</span> <span class="sc">-</span> h2) <span class="sc">*</span> beta<span class="sc">^</span><span class="dv">2</span> <span class="co"># Variance of the error term epsilon, ensuring desired h2</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>sigX <span class="ot">=</span> <span class="fu">sqrt</span>(sig2X)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>sigepsi <span class="ot">=</span> <span class="fu">sqrt</span>(sig2epsi)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Variance of X:"</span>, sig2X))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Variance of X: 0.1"</code></pre>
</div>
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Variance of Epsilon:"</span>, sig2epsi))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Variance of Epsilon: 3.6"</code></pre>
</div>
</div>
</section>
<section id="generate-single-instance-of-data" class="level2">
<h2 class="anchored" data-anchor-id="generate-single-instance-of-data">Generate Single Instance of Data</h2>
<p>Simulate vectors X, epsilon (error), and Ynull (outcome under null). Then calculate Yalt.</p>
<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Simulate predictor X and error term epsi</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>X <span class="ot">=</span> <span class="fu">rnorm</span>(nsamp, <span class="at">mean=</span><span class="dv">0</span>, <span class="at">sd=</span> sigX)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>epsi <span class="ot">=</span> <span class="fu">rnorm</span>(nsamp, <span class="at">mean=</span><span class="dv">0</span>, <span class="at">sd=</span>sigepsi)</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate Ynull (independent of X, here just sampling with variance related to beta for comparison)</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>Ynull <span class="ot">=</span> <span class="fu">rnorm</span>(nsamp, <span class="at">mean=</span><span class="dv">0</span>, <span class="at">sd=</span>beta) </span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate Yalt = X * beta + epsi</span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>Yalt <span class="ot">=</span> X <span class="sc">*</span> beta <span class="sc">+</span> epsi</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="visualize-the-data" class="level2">
<h2 class="anchored" data-anchor-id="visualize-the-data">Visualize the Data</h2>
<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize Data</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(X, Yalt, <span class="at">main=</span><span class="st">"Yalt vs X"</span>); <span class="fu">grid</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index-with-code_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(X, Ynull, <span class="at">main=</span><span class="st">"Ynull vs X"</span>); <span class="fu">grid</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index-with-code_files/figure-html/unnamed-chunk-3-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="test-associations-single-instance" class="level2">
<h2 class="anchored" data-anchor-id="test-associations-single-instance">Test Associations (Single Instance)</h2>
<p>Test the association between Ynull and X using linear regression.</p>
<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">lm</span>(Ynull <span class="sc">~</span> X))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = Ynull ~ X)

Residuals:
    Min      1Q  Median      3Q     Max 
-4.3548 -1.1589  0.0782  1.3735  4.6291 

Coefficients:
            Estimate Std. Error t value Pr(&gt;|t|)
(Intercept)  -0.0395     0.1951  -0.202    0.840
X             1.0095     0.6240   1.618    0.109

Residual standard error: 1.95 on 98 degrees of freedom
Multiple R-squared:  0.02601,   Adjusted R-squared:  0.01607 
F-statistic: 2.617 on 1 and 98 DF,  p-value: 0.1089</code></pre>
</div>
</div>
<p>Question: What is the p-value of the association between Ynull and X? Is it significant at the α=0.05 level?</p>
<p>Now test the association between Yalt and X.</p>
<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">lm</span>(Yalt <span class="sc">~</span> X))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = Yalt ~ X)

Residuals:
    Min      1Q  Median      3Q     Max 
-4.5453 -1.4809  0.1231  1.2189  4.1808 

Coefficients:
            Estimate Std. Error t value Pr(&gt;|t|)   
(Intercept)  -0.4244     0.1891  -2.244   0.0271 * 
X             2.0521     0.6050   3.392   0.0010 **
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 1.89 on 98 degrees of freedom
Multiple R-squared:  0.1051,    Adjusted R-squared:  0.09595 
F-statistic: 11.51 on 1 and 98 DF,  p-value: 0.001001</code></pre>
</div>
</div>
<p>Question: What is the p-value of the association between Yalt and X? Is it significant at the α=0.05 level?</p>
</section>
<section id="calculate-the-empirical-distribution-of-p-values" class="level2">
<h2 class="anchored" data-anchor-id="calculate-the-empirical-distribution-of-p-values">Calculate the Empirical Distribution of P-values</h2>
<p>To understand the distribution of p-values, we need to repeat the simulation many times.</p>
<section id="convenience-function-fastlm" class="level3">
<h3 class="anchored" data-anchor-id="convenience-function-fastlm">Convenience Function fastlm</h3>
<p>We’ll run 10,000 regressions. A simplified function fastlm can speed this up compared to repeated calls to lm().</p>
<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>fastlm <span class="ot">=</span> <span class="cf">function</span>(xx,yy) {</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>  <span class="do">## compute betahat (regression coef) and pvalue with Ftest</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>  <span class="do">## for now it does not take covariates</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>  df1 <span class="ot">=</span> <span class="dv">2</span> <span class="co"># DFs for model with predictor</span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>  df0 <span class="ot">=</span> <span class="dv">1</span> <span class="co"># DFs for null model (intercept only)</span></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>  ind <span class="ot">=</span> <span class="sc">!</span><span class="fu">is.na</span>(xx) <span class="sc">&amp;</span> <span class="sc">!</span><span class="fu">is.na</span>(yy)</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>  xx <span class="ot">=</span> xx[ind]</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>  yy <span class="ot">=</span> yy[ind]</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>  n <span class="ot">=</span> <span class="fu">sum</span>(ind)</span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>  xbar <span class="ot">=</span> <span class="fu">mean</span>(xx)</span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>  ybar <span class="ot">=</span> <span class="fu">mean</span>(yy)</span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a>  xx <span class="ot">=</span> xx <span class="sc">-</span> xbar <span class="co"># Center X</span></span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>  yy <span class="ot">=</span> yy <span class="sc">-</span> ybar <span class="co"># Center Y</span></span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a>  SXX <span class="ot">=</span> <span class="fu">sum</span>( xx<span class="sc">^</span><span class="dv">2</span> )</span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a>  SYY <span class="ot">=</span> <span class="fu">sum</span>( yy<span class="sc">^</span><span class="dv">2</span> )</span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true" tabindex="-1"></a>  SXY <span class="ot">=</span> <span class="fu">sum</span>( xx <span class="sc">*</span> yy )</span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb12-20"><a href="#cb12-20" aria-hidden="true" tabindex="-1"></a>  betahat <span class="ot">=</span> SXY <span class="sc">/</span> SXX</span>
<span id="cb12-21"><a href="#cb12-21" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb12-22"><a href="#cb12-22" aria-hidden="true" tabindex="-1"></a>  <span class="co"># RSS1 = Residual Sum of Squares for model Y ~ X</span></span>
<span id="cb12-23"><a href="#cb12-23" aria-hidden="true" tabindex="-1"></a>  RSS1 <span class="ot">=</span> <span class="fu">sum</span>( ( yy <span class="sc">-</span> xx <span class="sc">*</span> betahat )<span class="sc">^</span><span class="dv">2</span> ) </span>
<span id="cb12-24"><a href="#cb12-24" aria-hidden="true" tabindex="-1"></a>  <span class="co"># RSS0 = Residual Sum of Squares for model Y ~ 1 (equivalent to total sum of squares of centered Y)</span></span>
<span id="cb12-25"><a href="#cb12-25" aria-hidden="true" tabindex="-1"></a>  RSS0 <span class="ot">=</span> SYY </span>
<span id="cb12-26"><a href="#cb12-26" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb12-27"><a href="#cb12-27" aria-hidden="true" tabindex="-1"></a>  <span class="co"># F-statistic comparing the two models</span></span>
<span id="cb12-28"><a href="#cb12-28" aria-hidden="true" tabindex="-1"></a>  fstat <span class="ot">=</span> ( ( RSS0 <span class="sc">-</span> RSS1 ) <span class="sc">/</span> ( df1 <span class="sc">-</span> df0 ) )  <span class="sc">/</span> ( RSS1 <span class="sc">/</span> ( n <span class="sc">-</span> df1 ) ) </span>
<span id="cb12-29"><a href="#cb12-29" aria-hidden="true" tabindex="-1"></a>  <span class="co"># P-value from F-distribution</span></span>
<span id="cb12-30"><a href="#cb12-30" aria-hidden="true" tabindex="-1"></a>  pval <span class="ot">=</span> <span class="dv">1</span> <span class="sc">-</span> <span class="fu">pf</span>(fstat, <span class="at">df1 =</span> ( df1 <span class="sc">-</span> df0 ), <span class="at">df2 =</span> ( n <span class="sc">-</span> df1 )) </span>
<span id="cb12-31"><a href="#cb12-31" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb12-32"><a href="#cb12-32" aria-hidden="true" tabindex="-1"></a>  res <span class="ot">=</span> <span class="fu">list</span>(<span class="at">betahat =</span> betahat, <span class="at">pval =</span> pval)</span>
<span id="cb12-33"><a href="#cb12-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(res)</span>
<span id="cb12-34"><a href="#cb12-34" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="run-10000-simulations" class="level3">
<h3 class="anchored" data-anchor-id="run-10000-simulations">Run 10,000 Simulations</h3>
<p>Simulate X, Ynull, and Yalt 10,000 times. Store X and the outcomes in matrices.</p>
<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>nsim <span class="ot">=</span> <span class="dv">10000</span> <span class="co"># Number of simulations</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Simulate matrices: rows=samples, cols=simulations</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>Xmat <span class="ot">=</span> <span class="fu">matrix</span>(<span class="fu">rnorm</span>(nsim <span class="sc">*</span> nsamp, <span class="at">mean=</span><span class="dv">0</span>, <span class="at">sd=</span> sigX), nsamp, nsim)</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>epsimat <span class="ot">=</span> <span class="fu">matrix</span>(<span class="fu">rnorm</span>(nsim <span class="sc">*</span> nsamp, <span class="at">mean=</span><span class="dv">0</span>, <span class="at">sd=</span>sigepsi), nsamp, nsim)</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate Y matrices based on the formulas</span></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>Ymat_alt <span class="ot">=</span> Xmat <span class="sc">*</span> beta <span class="sc">+</span> epsimat </span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>Ymat_null <span class="ot">=</span> <span class="fu">matrix</span>(<span class="fu">rnorm</span>(nsim <span class="sc">*</span> nsamp, <span class="at">mean=</span><span class="dv">0</span>, <span class="at">sd=</span>beta), nsamp, nsim) <span class="co"># Null Ys</span></span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Check dimensions</span></span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(Ymat_null)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1]   100 10000</code></pre>
</div>
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(Ymat_alt)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1]   100 10000</code></pre>
</div>
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Assign column names for reference</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(Ymat_null) <span class="ot">=</span> <span class="fu">paste0</span>(<span class="st">"sim"</span>, <span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(Ymat_null))</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(Ymat_alt) <span class="ot">=</span> <span class="fu">colnames</span>(Ymat_null)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="calculate-p-values-under-the-null" class="level3">
<h3 class="anchored" data-anchor-id="calculate-p-values-under-the-null">Calculate P-values Under the Null</h3>
<p>Run regressions for Ymat_null ~ Xmat for each simulation and store the p-values and coefficients.</p>
<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>pvec_null <span class="ot">=</span> <span class="fu">rep</span>(<span class="cn">NA</span>, nsim)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>bvec_null <span class="ot">=</span> <span class="fu">rep</span>(<span class="cn">NA</span>, nsim)</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(ss <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>nsim) {</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>  fit <span class="ot">=</span> <span class="fu">fastlm</span>(Xmat[,ss], Ymat_null[,ss])</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>  pvec_null[ss] <span class="ot">=</span> fit<span class="sc">$</span>pval  </span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>  bvec_null[ss] <span class="ot">=</span> fit<span class="sc">$</span>betahat</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(pvec_null)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>     Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
0.0000021 0.2438476 0.4903663 0.4935828 0.7401156 0.9998408 </code></pre>
</div>
</div>
<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(pvec_null, <span class="at">xlab=</span><span class="st">"p-value"</span>, <span class="at">main=</span><span class="st">"Histogram of p-values under Null"</span>, <span class="at">breaks=</span><span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index-with-code_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Questions: - The histogram should look approximately uniform. Why? - How many simulations under the null yield p-value below 0.05? What percentage is that?</p>
<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(pvec_null <span class="sc">&lt;</span> <span class="fl">0.05</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 525</code></pre>
</div>
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(pvec_null <span class="sc">&lt;</span> <span class="fl">0.05</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.0525</code></pre>
</div>
</div>
<p>What proportion of simulations do you expect to have p-values &lt; α (for any α between 0 and 1) under the null? Why does this uniform distribution highlight the need for correction when performing many tests?</p>
</section>
</section>
</section>
<section id="bonferroni-correction" class="level1">
<h1>Bonferroni Correction</h1>
<p>The simplest correction is the Bonferroni correction.</p>
<p>Method: Divide the desired significance level (e.g., 0.05) by the total number of tests performed (m, here nsim). Use this adjusted value as the new significance threshold.</p>
<p>Goal: Controls the Family-Wise Error Rate (FWER) - the probability of calling one or more significant results across all tests under the null. P(FP&gt;=1) ≤ α.</p>
<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>BF_thres <span class="ot">=</span> <span class="fl">0.05</span> <span class="sc">/</span> nsim <span class="co"># Bonferroni significance threshold</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Bonferroni threshold:"</span>, BF_thres))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Bonferroni threshold: 5e-06"</code></pre>
</div>
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Number of Bonferroni significant associations under the NULL</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Found under Null:"</span>, <span class="fu">sum</span>(pvec_null <span class="sc">&lt;</span> BF_thres)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Found under Null: 1"</code></pre>
</div>
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Proportion under Null:"</span>, <span class="fu">mean</span>(pvec_null <span class="sc">&lt;</span> BF_thres)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Proportion under Null: 1e-04"</code></pre>
</div>
</div>
<p>Questions: - What is the Bonferroni threshold for significance in this simulation? - How many false positives (significant results under the null) did we find using this strict threshold?</p>
</section>
<section id="mix-of-null-and-alternative-hypotheses" class="level1">
<h1>Mix of Null and Alternative Hypotheses</h1>
<p>Real-world datasets (like GWAS) contain a mixture of true associations (alternative hypothesis) and null results. Let’s simulate this.</p>
<section id="create-mixed-data" class="level2">
<h2 class="anchored" data-anchor-id="create-mixed-data">Create Mixed Data</h2>
<p>Assume a certain proportion (prop_alt) of our simulations represent true associations.</p>
<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>prop_alt <span class="ot">=</span> <span class="fl">0.20</span> <span class="co"># Define proportion of alternative Ys in the mixture</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a selection vector: 1 if alternative, 0 if null</span></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>selectvec <span class="ot">=</span> <span class="fu">rbinom</span>(nsim, <span class="dv">1</span>, prop_alt) </span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(selectvec) <span class="ot">=</span> <span class="fu">colnames</span>(Ymat_alt) <span class="co"># Keep track</span></span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(selectvec) <span class="co"># Show counts of null (0) vs alternative (1)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>selectvec
   0    1 
7996 2004 </code></pre>
</div>
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create the mixed Y matrix</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="co"># If selectvec[i] is 1, use Ymat_alt[,i]; if 0, use Ymat_null[,i]</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>Ymat_mix <span class="ot">=</span> <span class="fu">sweep</span>(Ymat_alt, <span class="dv">2</span>, selectvec, <span class="at">FUN=</span><span class="st">'*'</span>) <span class="sc">+</span> <span class="fu">sweep</span>(Ymat_null, <span class="dv">2</span>, <span class="dv">1</span> <span class="sc">-</span> selectvec, <span class="at">FUN=</span><span class="st">'*'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="calculate-p-values-for-mixed-data" class="level2">
<h2 class="anchored" data-anchor-id="calculate-p-values-for-mixed-data">Calculate P-values for Mixed Data</h2>
<p>Run regressions for Ymat_mix ~ Xmat.</p>
<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>pvec_mix <span class="ot">=</span> <span class="fu">rep</span>(<span class="cn">NA</span>, nsim)</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>bvec_mix <span class="ot">=</span> <span class="fu">rep</span>(<span class="cn">NA</span>, nsim)</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(ss <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>nsim) {</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>  fit <span class="ot">=</span> <span class="fu">fastlm</span>(Xmat[,ss], Ymat_mix[,ss])</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>  pvec_mix[ss] <span class="ot">=</span> fit<span class="sc">$</span>pval  </span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>  bvec_mix[ss] <span class="ot">=</span> fit<span class="sc">$</span>betahat</span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(pvec_mix)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
0.00000 0.07854 0.37313 0.40036 0.67907 0.99984 </code></pre>
</div>
</div>
<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(pvec_mix, <span class="at">xlab=</span><span class="st">"p-value"</span>, <span class="at">main=</span><span class="st">"Histogram of p-values under mixture"</span>, <span class="at">breaks=</span><span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index-with-code_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="interpreting-mixed-p-values" class="level2">
<h2 class="anchored" data-anchor-id="interpreting-mixed-p-values">Interpreting Mixed P-values</h2>
<p>The histogram shows a spike near zero (from true alternatives) superimposed on a uniform background (from true nulls).</p>
<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>thres <span class="ot">=</span> <span class="fl">0.05</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>m_signif <span class="ot">=</span> <span class="fu">sum</span>(pvec_mix <span class="sc">&lt;</span> thres)      <span class="co"># Observed number of significant associations</span></span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>m_expected_null <span class="ot">=</span> thres <span class="sc">*</span> <span class="fu">sum</span>(selectvec <span class="sc">==</span> <span class="dv">0</span>) <span class="co"># Expected number of FPs from nulls</span></span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>m_expected_all_null <span class="ot">=</span> thres <span class="sc">*</span> nsim <span class="co"># Expected if *all* were null</span></span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Observed significant (p&lt;0.05):"</span>, m_signif))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Observed significant (p&lt;0.05): 2206"</code></pre>
</div>
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Expected false positives if all were null:"</span>, m_expected_all_null))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Expected false positives if all were null: 500"</code></pre>
</div>
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Expected false positives from the actual nulls:"</span>, <span class="fu">round</span>(m_expected_null)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Expected false positives from the actual nulls: 400"</code></pre>
</div>
</div>
<p>We observed many more significant results than expected if all tests were null. How can we estimate the proportion of false discoveries among those we call significant?</p>
</section>
<section id="false-discovery-rate-fdr-estimate-using-known-truth" class="level2">
<h2 class="anchored" data-anchor-id="false-discovery-rate-fdr-estimate-using-known-truth">False Discovery Rate (FDR) Estimate (using known truth)</h2>
<p>Since this is a simulation, we know which tests were truly null (selectvec == 0). We can calculate the actual FDR for a given p-value threshold.</p>
<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="co"># FDR = (Number of significant tests that are truly null) / (Total number of significant tests)</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>FP <span class="ot">=</span> <span class="fu">sum</span>(pvec_mix <span class="sc">&lt;</span> thres <span class="sc">&amp;</span> selectvec <span class="sc">==</span> <span class="dv">0</span>) <span class="co"># False Positives</span></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>TP <span class="ot">=</span> <span class="fu">sum</span>(pvec_mix <span class="sc">&lt;</span> thres <span class="sc">&amp;</span> selectvec <span class="sc">==</span> <span class="dv">1</span>) <span class="co"># True Positives</span></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>S <span class="ot">=</span> <span class="fu">sum</span>(pvec_mix <span class="sc">&lt;</span> thres) <span class="co"># Total Significant (S = FP + TP)</span></span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a>FDR_sim <span class="ot">=</span> FP <span class="sc">/</span> S </span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Simulated FDR at p &lt;"</span>, thres, <span class="st">":"</span>, <span class="fu">round</span>(FDR_sim, <span class="dv">4</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Simulated FDR at p &lt; 0.05 : 0.1886"</code></pre>
</div>
</div>
<p>This means that if we used p &lt; 0.05 as our criterion, about round(FDR_sim*100, 1)% of our significant findings would actually be false positives.</p>
<p>Questions: - What is the proportion of false discoveries (FDR) if we use a significance level of 0.01?</p>
<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>thres_01 <span class="ot">=</span> <span class="fl">0.01</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>FP_01 <span class="ot">=</span> <span class="fu">sum</span>(pvec_mix <span class="sc">&lt;</span> thres_01 <span class="sc">&amp;</span> selectvec <span class="sc">==</span> <span class="dv">0</span>)</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>S_01 <span class="ot">=</span> <span class="fu">sum</span>(pvec_mix <span class="sc">&lt;</span> thres_01)</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>FDR_sim_01 <span class="ot">=</span> FP_01 <span class="sc">/</span> S_01</span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Simulated FDR at p &lt; 0.01:"</span>, <span class="fu">round</span>(FDR_sim_01, <span class="dv">4</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Simulated FDR at p &lt; 0.01: 0.0582"</code></pre>
</div>
</div>
<ul>
<li>What is the proportion of false discoveries if we use the Bonferroni threshold?</li>
</ul>
<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>FP_bf <span class="ot">=</span> <span class="fu">sum</span>(pvec_mix <span class="sc">&lt;</span> BF_thres <span class="sc">&amp;</span> selectvec <span class="sc">==</span> <span class="dv">0</span>)</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>S_bf <span class="ot">=</span> <span class="fu">sum</span>(pvec_mix <span class="sc">&lt;</span> BF_thres)</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>FDR_sim_bf <span class="ot">=</span> <span class="fu">ifelse</span>(S_bf <span class="sc">&gt;</span> <span class="dv">0</span>, FP_bf <span class="sc">/</span> S_bf, <span class="dv">0</span>) <span class="co"># Avoid division by zero</span></span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Simulated FDR at Bonferroni p &lt;"</span>, BF_thres, <span class="st">":"</span>, <span class="fu">round</span>(FDR_sim_bf, <span class="dv">4</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Simulated FDR at Bonferroni p &lt; 5e-06 : 0.0055"</code></pre>
</div>
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Number significant at Bonferroni:"</span>, S_bf))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Number significant at Bonferroni: 181"</code></pre>
</div>
</div>
<ul>
<li>What is the proportion of missed signals (False Negative Rate among true alternatives) using the Bonferroni threshold?</li>
</ul>
<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>FN_bf <span class="ot">=</span> <span class="fu">sum</span>(pvec_mix <span class="sc">&gt;=</span> BF_thres <span class="sc">&amp;</span> selectvec <span class="sc">==</span> <span class="dv">1</span>) <span class="co"># False Negatives</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>Total_Alt <span class="ot">=</span> <span class="fu">sum</span>(selectvec <span class="sc">==</span> <span class="dv">1</span>) <span class="co"># Total true alternatives</span></span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>FNR_sim_bf <span class="ot">=</span> FN_bf <span class="sc">/</span> Total_Alt</span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Simulated FNR at Bonferroni p &lt;"</span>, BF_thres, <span class="st">":"</span>, <span class="fu">round</span>(FNR_sim_bf, <span class="dv">4</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Simulated FNR at Bonferroni p &lt; 5e-06 : 0.9102"</code></pre>
</div>
</div>
</section>
</section>
<section id="common-approaches-to-control-type-i-errors" class="level1">
<h1>Common Approaches to Control Type I Errors</h1>
<p>Let m be the total number of tests. Consider the following table:</p>
<table class="caption-top table">
<thead>
<tr class="header">
<th></th>
<th>Called Not Significant</th>
<th>Called Significant</th>
<th>Total</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Null True</td>
<td>TN</td>
<td>FP</td>
<td><span class="math inline">\(m_0\)</span></td>
</tr>
<tr class="even">
<td>Alternative True</td>
<td>FN</td>
<td>TP</td>
<td><span class="math inline">\(m_1\)</span></td>
</tr>
<tr class="odd">
<td>Total</td>
<td><span class="math inline">\(m\)</span> - S</td>
<td>S</td>
<td><span class="math inline">\(m\)</span></td>
</tr>
</tbody>
</table>
<p>(m total # tests, <span class="math inline">\(m_0\)</span> # null tests, FP false positives, TP true positives, TN true negatives, FN false negatives, S total significant)</p>
<ul>
<li>FWER (Family-Wise Error Rate): P(FP ≥ 1). Probability of at least one false positive. Controlled by Bonferroni. Often too strict.</li>
<li>FDR (False Discovery Rate): E[FP / S | S &gt; 0]. Expected proportion of false positives among all significant results. A more liberal approach, often preferred when many tests are performed and finding true positives is important.</li>
</ul>
<section id="table-of-simulation-results" class="level2">
<h2 class="anchored" data-anchor-id="table-of-simulation-results">Table of Simulation Results</h2>
<p>Let’s construct the confusion table for our mixed simulation using alpha = 0.05.</p>
<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>count_table <span class="ot">=</span> <span class="fu">t</span>(<span class="fu">table</span>(<span class="at">Significant =</span> pvec_mix <span class="sc">&lt;</span> <span class="fl">0.05</span>, <span class="at">True_Status =</span> selectvec))</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Reorder and rename for standard confusion matrix format</span></span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Rows: True Status (0=Null, 1=Alt); Columns: Called Status (FALSE=Not Sig, TRUE=Sig)</span></span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a>conf_matrix <span class="ot">=</span> count_table[, <span class="fu">c</span>(<span class="st">"FALSE"</span>, <span class="st">"TRUE"</span>)] </span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(conf_matrix) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"Called Not Significant"</span>, <span class="st">"Called Significant"</span>)</span>
<span id="cb53-6"><a href="#cb53-6" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(conf_matrix) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"Null True"</span>, <span class="st">"Alt True"</span>)</span>
<span id="cb53-7"><a href="#cb53-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-8"><a href="#cb53-8" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(conf_matrix, <span class="at">caption=</span><span class="st">"Confusion Matrix at p &lt; 0.05"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<table class="caption-top table table-sm table-striped small">
<caption>Confusion Matrix at p &lt; 0.05</caption>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">Called Not Significant</th>
<th style="text-align: right;">Called Significant</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Null True</td>
<td style="text-align: right;">7580</td>
<td style="text-align: right;">416</td>
</tr>
<tr class="even">
<td style="text-align: left;">Alt True</td>
<td style="text-align: right;">214</td>
<td style="text-align: right;">1790</td>
</tr>
</tbody>
</table>
</div>
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extracting values</span></span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>TN <span class="ot">=</span> conf_matrix[<span class="st">"Null True"</span>, <span class="st">"Called Not Significant"</span>]</span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a>FP <span class="ot">=</span> conf_matrix[<span class="st">"Null True"</span>, <span class="st">"Called Significant"</span>]</span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a>FN <span class="ot">=</span> conf_matrix[<span class="st">"Alt True"</span>, <span class="st">"Called Not Significant"</span>]</span>
<span id="cb54-5"><a href="#cb54-5" aria-hidden="true" tabindex="-1"></a>TP <span class="ot">=</span> conf_matrix[<span class="st">"Alt True"</span>, <span class="st">"Called Significant"</span>]</span>
<span id="cb54-6"><a href="#cb54-6" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"TN:"</span>, TN, <span class="st">" FP:"</span>, FP, <span class="st">" FN:"</span>, FN, <span class="st">" TP:"</span>, TP))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "TN: 7580  FP: 416  FN: 214  TP: 1790"</code></pre>
</div>
</div>
</section>
</section>
<section id="use-qvalue-package-to-estimate-fdr" class="level1">
<h1>Use qvalue Package to Estimate FDR</h1>
<p>In real analysis, we don’t know selectvec. The qvalue package estimates FDR and related quantities directly from the p-value distribution.</p>
<p>q-value: The minimum FDR at which a test would be called significant. If you call all tests with q ≤ x significant, you expect the FDR to be approximately x.</p>
<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Install qvalue package if you haven't already</span></span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a><span class="co"># if (!requireNamespace("BiocManager", quietly = TRUE))</span></span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a><span class="co">#    install.packages("BiocManager")</span></span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a><span class="co">#BiocManager::install("qvalue")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(qvalue)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate q-values for the mixed p-values</span></span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>qres_mix <span class="ot">=</span> <span class="fu">qvalue</span>(pvec_mix)</span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>qvec_mix <span class="ot">=</span> qres_mix<span class="sc">$</span>qvalues</span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-5"><a href="#cb58-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Also calculate for the null p-values for comparison</span></span>
<span id="cb58-6"><a href="#cb58-6" aria-hidden="true" tabindex="-1"></a>qres_null <span class="ot">=</span> <span class="fu">qvalue</span>(pvec_null)</span>
<span id="cb58-7"><a href="#cb58-7" aria-hidden="true" tabindex="-1"></a>qvec_null <span class="ot">=</span> qres_null<span class="sc">$</span>qvalues</span>
<span id="cb58-8"><a href="#cb58-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-9"><a href="#cb58-9" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(qvec_mix)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>     Min.   1st Qu.    Median      Mean   3rd Qu.      Max. 
0.0000019 0.2391430 0.5681539 0.4635186 0.6894836 0.7615203 </code></pre>
</div>
</div>
<section id="visualize-q-values" class="level2">
<h2 class="anchored" data-anchor-id="visualize-q-values">Visualize Q-values</h2>
<p>Let’s see if small q-values correspond to true associations (selectvec = 1).</p>
<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(qvec_mix <span class="sc">~</span> selectvec, <span class="at">main=</span><span class="st">'Q-value by True Status'</span>, <span class="at">xlab=</span><span class="st">"True Status (0=Null, 1=Alt)"</span>, <span class="at">ylab=</span><span class="st">"Q-value"</span>)</span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a><span class="fu">grid</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index-with-code_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Plot sorted q-values, colored by true status.</p>
<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>ind <span class="ot">=</span> <span class="fu">order</span>(qvec_mix, <span class="at">decreasing=</span><span class="cn">FALSE</span>)</span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Using selectvec*2 + 1 maps 0-&gt;1 (black) and 1-&gt;3 (green) for plotting colors</span></span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">sort</span>(qvec_mix), <span class="at">col=</span>selectvec[ind]<span class="sc">*</span><span class="dv">2</span> <span class="sc">+</span> <span class="dv">1</span>, </span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">pch=</span><span class="dv">16</span>, <span class="co">#selectvec[ind] + 1, </span></span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a>     <span class="co"># lwd=selectvec[ind]*2 + 1,</span></span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a>     <span class="at">main=</span><span class="st">"Sorted Q-values by True Status"</span>,</span>
<span id="cb61-7"><a href="#cb61-7" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab=</span><span class="st">"Tests ordered by Q-value"</span>, <span class="at">ylab=</span><span class="st">"Q-value"</span>)</span>
<span id="cb61-8"><a href="#cb61-8" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"bottomright"</span>, <span class="at">legend=</span><span class="fu">c</span>(<span class="st">"True Null"</span>, <span class="st">"True Alternative"</span>), <span class="at">pch=</span><span class="dv">16</span>, <span class="at">col=</span><span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index-with-code_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Questions: - Do small q-values tend to correspond to true alternatives (green points)? - Interpret the sorted q-value plot: What does the pattern show?</p>
<p>Compare p-value and q-value distributions by true status.</p>
<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(pvec_mix <span class="sc">~</span> selectvec, <span class="at">main=</span><span class="st">'P-value by True Status'</span>, <span class="at">xlab=</span><span class="st">"True Status (0=Null, 1=Alt)"</span>, <span class="at">ylab=</span><span class="st">"P-value"</span>); <span class="fu">grid</span>()</span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(qvec_mix <span class="sc">~</span> selectvec, <span class="at">main=</span><span class="st">'Q-value by True Status'</span>, <span class="at">xlab=</span><span class="st">"True Status (0=Null, 1=Alt)"</span>, <span class="at">ylab=</span><span class="st">"Q-value"</span>); <span class="fu">grid</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index-with-code_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Question: Interpret these boxplots. How do q-values differ from p-values, especially for the null tests?</p>
</section>
<section id="how-do-q-values-and-p-values-relate" class="level2">
<h2 class="anchored" data-anchor-id="how-do-q-values-and-p-values-relate">How Do Q-values and P-values Relate?</h2>
<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pvec_null, qvec_null, <span class="at">main=</span><span class="st">'Q-value vs P-value (All Null)'</span>, <span class="at">xlab=</span><span class="st">"P-value"</span>, <span class="at">ylab=</span><span class="st">"Q-value"</span>); <span class="fu">grid</span>(); <span class="fu">abline</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="at">col=</span><span class="st">'red'</span>,<span class="at">lty=</span><span class="dv">2</span>)</span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pvec_mix, qvec_mix, <span class="at">main=</span><span class="st">'Q-value vs P-value (Mixture)'</span>, <span class="at">xlab=</span><span class="st">"P-value"</span>, <span class="at">ylab=</span><span class="st">"Q-value"</span>); <span class="fu">grid</span>(); <span class="fu">abline</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="at">col=</span><span class="st">'red'</span>,<span class="at">lty=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="index-with-code_files/figure-html/unnamed-chunk-27-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Q-values are monotonically increasing functions of p-values. Note that q-values are often ≥ p-values. The q-value can be interpreted as the FDR associated with calling that specific test significant.</p>
<p>Questions: - What is the smallest q-value when all simulations are from the null? Why? - What is the smallest q-value when simulations are from the mixture?</p>
</section>
<section id="estimating-π₀-proportion-of-true-nulls" class="level2">
<h2 class="anchored" data-anchor-id="estimating-π₀-proportion-of-true-nulls">Estimating π₀ (Proportion of True Nulls)</h2>
<p>The qvalue package also estimates π₀, the overall proportion of features (tests) that are truly null.</p>
<p>π₁ = 1 - π₀ is the proportion of true alternative hypotheses.</p>
<div class="cell">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Estimated pi0 when all tests were truly null</span></span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Estimated pi0 (all null):"</span>, <span class="fu">round</span>(qres_null<span class="sc">$</span>pi0, <span class="dv">4</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Estimated pi0 (all null): 0.9582"</code></pre>
</div>
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Estimated pi0 for the mixed data</span></span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Estimated pi0 (mixture):"</span>, <span class="fu">round</span>(qres_mix<span class="sc">$</span>pi0, <span class="dv">4</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Estimated pi0 (mixture): 0.7616"</code></pre>
</div>
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="co"># True proportion used in simulation</span></span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"True pi0 used:"</span>, <span class="dv">1</span> <span class="sc">-</span> prop_alt))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "True pi0 used: 0.8"</code></pre>
</div>
</div>
<p>Question: How well did the qvalue package estimate the true proportion of null hypotheses (pi0) in our mixed simulation?</p>
</section>
</section>
<section id="references" class="level1">
<h1>References</h1>
<p>Storey, John D., and Robert Tibshirani. 2003. “Statistical Significance for Genomewide Studies.” Proceedings of the National Academy of Sciences 100 (16): 9440–45.</p>


<!-- -->

</section>

<p>© <a href="https://hakyimlab.org">HakyImLab and Listed Authors</a> - <a href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0 License</a></p></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb72" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> lecture 4 - multiple testing correction - with code</span></span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a><span class="an">author:</span><span class="co"> "Haky Im"</span></span>
<span id="cb72-4"><a href="#cb72-4" aria-hidden="true" tabindex="-1"></a><span class="an">date:</span><span class="co"> "March 31, 2025"</span></span>
<span id="cb72-5"><a href="#cb72-5" aria-hidden="true" tabindex="-1"></a><span class="an">description:</span><span class="co"> "Lecture 4"</span></span>
<span id="cb72-6"><a href="#cb72-6" aria-hidden="true" tabindex="-1"></a><span class="an">categories:</span></span>
<span id="cb72-7"><a href="#cb72-7" aria-hidden="true" tabindex="-1"></a><span class="co">  - bios25328</span></span>
<span id="cb72-8"><a href="#cb72-8" aria-hidden="true" tabindex="-1"></a><span class="co">  - notebook</span></span>
<span id="cb72-9"><a href="#cb72-9" aria-hidden="true" tabindex="-1"></a><span class="an">freeze:</span><span class="co"> true</span></span>
<span id="cb72-10"><a href="#cb72-10" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb72-11"><a href="#cb72-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-12"><a href="#cb72-12" aria-hidden="true" tabindex="-1"></a>Find the lecture notes <span class="co">[</span><span class="ot">here</span><span class="co">](https://www.icloud.com/keynote/07fVVMHEvj8f2aG5zyyFfXisA#L4-Multiple-Testing-Correction-2025)</span>.</span>
<span id="cb72-13"><a href="#cb72-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-14"><a href="#cb72-14" aria-hidden="true" tabindex="-1"></a><span class="in">```{r setup, include=FALSE}</span></span>
<span id="cb72-15"><a href="#cb72-15" aria-hidden="true" tabindex="-1"></a><span class="in"># Ensure necessary packages are loaded silently if needed later</span></span>
<span id="cb72-16"><a href="#cb72-16" aria-hidden="true" tabindex="-1"></a><span class="in"># library(qvalue) </span></span>
<span id="cb72-17"><a href="#cb72-17" aria-hidden="true" tabindex="-1"></a><span class="in"># library(knitr)</span></span>
<span id="cb72-18"><a href="#cb72-18" aria-hidden="true" tabindex="-1"></a><span class="in"># Set seed for reproducibility if desired globally</span></span>
<span id="cb72-19"><a href="#cb72-19" aria-hidden="true" tabindex="-1"></a><span class="in">set.seed(20210108) </span></span>
<span id="cb72-20"><a href="#cb72-20" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb72-21"><a href="#cb72-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-22"><a href="#cb72-22" aria-hidden="true" tabindex="-1"></a><span class="fu"># Learning Objectives</span></span>
<span id="cb72-23"><a href="#cb72-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-24"><a href="#cb72-24" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Build intuition about p-values when multiple testing is performed via simulations.</span>
<span id="cb72-25"><a href="#cb72-25" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Recognize the need for multiple testing correction.</span>
<span id="cb72-26"><a href="#cb72-26" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Present methods to correct for multiple testing:</span>
<span id="cb72-27"><a href="#cb72-27" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span>Bonferroni correction</span>
<span id="cb72-28"><a href="#cb72-28" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span>FDR (false discovery rate)</span>
<span id="cb72-29"><a href="#cb72-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-30"><a href="#cb72-30" aria-hidden="true" tabindex="-1"></a><span class="fu"># Why Do We Need Multiple Testing Correction?</span></span>
<span id="cb72-31"><a href="#cb72-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-32"><a href="#cb72-32" aria-hidden="true" tabindex="-1"></a>When performing many statistical tests simultaneously, the probability of finding a "significant" result just by chance (a Type I error or false positive) increases substantially. This is famously illustrated by the XKCD comic on significance (link to XKCD #882). We need methods to control the overall error rate across all tests.</span>
<span id="cb72-33"><a href="#cb72-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-34"><a href="#cb72-34" aria-hidden="true" tabindex="-1"></a><span class="fu"># What Do P-values Look Like Under Null and Alternative Hypotheses?</span></span>
<span id="cb72-35"><a href="#cb72-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-36"><a href="#cb72-36" aria-hidden="true" tabindex="-1"></a><span class="fu">## Simulate Data: Null vs. Alternative</span></span>
<span id="cb72-37"><a href="#cb72-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-38"><a href="#cb72-38" aria-hidden="true" tabindex="-1"></a>We'll start by simulating data under two scenarios:</span>
<span id="cb72-39"><a href="#cb72-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-40"><a href="#cb72-40" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Null Hypothesis (Ynull): The outcome variable is independent of the predictor X.</span>
<span id="cb72-41"><a href="#cb72-41" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Alternative Hypothesis (Yalt): The outcome variable depends on the predictor X via the relationship Yalt = X * beta + epsilon.</span>
<span id="cb72-42"><a href="#cb72-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-43"><a href="#cb72-43" aria-hidden="true" tabindex="-1"></a><span class="fu">## Simulation Parameters</span></span>
<span id="cb72-44"><a href="#cb72-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-45"><a href="#cb72-45" aria-hidden="true" tabindex="-1"></a>First, define some parameters for the simulation.</span>
<span id="cb72-46"><a href="#cb72-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-49"><a href="#cb72-49" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb72-50"><a href="#cb72-50" aria-hidden="true" tabindex="-1"></a><span class="co"># Parameters</span></span>
<span id="cb72-51"><a href="#cb72-51" aria-hidden="true" tabindex="-1"></a>nsamp <span class="ot">=</span> <span class="dv">100</span>     <span class="co"># Number of samples</span></span>
<span id="cb72-52"><a href="#cb72-52" aria-hidden="true" tabindex="-1"></a>beta <span class="ot">=</span> <span class="dv">2</span>        <span class="co"># Effect size for the alternative hypothesis</span></span>
<span id="cb72-53"><a href="#cb72-53" aria-hidden="true" tabindex="-1"></a>h2 <span class="ot">=</span> <span class="fl">0.1</span>        <span class="co"># Proportion of variance in Yalt explained by X (related to signal strength)</span></span>
<span id="cb72-54"><a href="#cb72-54" aria-hidden="true" tabindex="-1"></a>sig2X <span class="ot">=</span> h2      <span class="co"># Variance of X (scaled for simplicity here)</span></span>
<span id="cb72-55"><a href="#cb72-55" aria-hidden="true" tabindex="-1"></a>sig2epsi <span class="ot">=</span> (<span class="dv">1</span> <span class="sc">-</span> h2) <span class="sc">*</span> beta<span class="sc">^</span><span class="dv">2</span> <span class="co"># Variance of the error term epsilon, ensuring desired h2</span></span>
<span id="cb72-56"><a href="#cb72-56" aria-hidden="true" tabindex="-1"></a>sigX <span class="ot">=</span> <span class="fu">sqrt</span>(sig2X)</span>
<span id="cb72-57"><a href="#cb72-57" aria-hidden="true" tabindex="-1"></a>sigepsi <span class="ot">=</span> <span class="fu">sqrt</span>(sig2epsi)</span>
<span id="cb72-58"><a href="#cb72-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-59"><a href="#cb72-59" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Variance of X:"</span>, sig2X))</span>
<span id="cb72-60"><a href="#cb72-60" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Variance of Epsilon:"</span>, sig2epsi))</span>
<span id="cb72-61"><a href="#cb72-61" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb72-62"><a href="#cb72-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-63"><a href="#cb72-63" aria-hidden="true" tabindex="-1"></a><span class="fu">## Generate Single Instance of Data</span></span>
<span id="cb72-64"><a href="#cb72-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-65"><a href="#cb72-65" aria-hidden="true" tabindex="-1"></a>Simulate vectors X, epsilon (error), and Ynull (outcome under null). Then calculate Yalt.</span>
<span id="cb72-66"><a href="#cb72-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-69"><a href="#cb72-69" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb72-70"><a href="#cb72-70" aria-hidden="true" tabindex="-1"></a><span class="co"># Simulate predictor X and error term epsi</span></span>
<span id="cb72-71"><a href="#cb72-71" aria-hidden="true" tabindex="-1"></a>X <span class="ot">=</span> <span class="fu">rnorm</span>(nsamp, <span class="at">mean=</span><span class="dv">0</span>, <span class="at">sd=</span> sigX)</span>
<span id="cb72-72"><a href="#cb72-72" aria-hidden="true" tabindex="-1"></a>epsi <span class="ot">=</span> <span class="fu">rnorm</span>(nsamp, <span class="at">mean=</span><span class="dv">0</span>, <span class="at">sd=</span>sigepsi)</span>
<span id="cb72-73"><a href="#cb72-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-74"><a href="#cb72-74" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate Ynull (independent of X, here just sampling with variance related to beta for comparison)</span></span>
<span id="cb72-75"><a href="#cb72-75" aria-hidden="true" tabindex="-1"></a>Ynull <span class="ot">=</span> <span class="fu">rnorm</span>(nsamp, <span class="at">mean=</span><span class="dv">0</span>, <span class="at">sd=</span>beta) </span>
<span id="cb72-76"><a href="#cb72-76" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-77"><a href="#cb72-77" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate Yalt = X * beta + epsi</span></span>
<span id="cb72-78"><a href="#cb72-78" aria-hidden="true" tabindex="-1"></a>Yalt <span class="ot">=</span> X <span class="sc">*</span> beta <span class="sc">+</span> epsi</span>
<span id="cb72-79"><a href="#cb72-79" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb72-80"><a href="#cb72-80" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-81"><a href="#cb72-81" aria-hidden="true" tabindex="-1"></a><span class="fu">## Visualize the Data</span></span>
<span id="cb72-82"><a href="#cb72-82" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-85"><a href="#cb72-85" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb72-86"><a href="#cb72-86" aria-hidden="true" tabindex="-1"></a><span class="co"># Visualize Data</span></span>
<span id="cb72-87"><a href="#cb72-87" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(X, Yalt, <span class="at">main=</span><span class="st">"Yalt vs X"</span>); <span class="fu">grid</span>()</span>
<span id="cb72-88"><a href="#cb72-88" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(X, Ynull, <span class="at">main=</span><span class="st">"Ynull vs X"</span>); <span class="fu">grid</span>()</span>
<span id="cb72-89"><a href="#cb72-89" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb72-90"><a href="#cb72-90" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-91"><a href="#cb72-91" aria-hidden="true" tabindex="-1"></a><span class="fu">## Test Associations (Single Instance)</span></span>
<span id="cb72-92"><a href="#cb72-92" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-93"><a href="#cb72-93" aria-hidden="true" tabindex="-1"></a>Test the association between Ynull and X using linear regression.</span>
<span id="cb72-94"><a href="#cb72-94" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-97"><a href="#cb72-97" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb72-98"><a href="#cb72-98" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">lm</span>(Ynull <span class="sc">~</span> X))</span>
<span id="cb72-99"><a href="#cb72-99" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb72-100"><a href="#cb72-100" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-101"><a href="#cb72-101" aria-hidden="true" tabindex="-1"></a>Question: What is the p-value of the association between Ynull and X? Is it significant at the α=0.05 level?</span>
<span id="cb72-102"><a href="#cb72-102" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-103"><a href="#cb72-103" aria-hidden="true" tabindex="-1"></a>Now test the association between Yalt and X.</span>
<span id="cb72-104"><a href="#cb72-104" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-107"><a href="#cb72-107" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb72-108"><a href="#cb72-108" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">lm</span>(Yalt <span class="sc">~</span> X))</span>
<span id="cb72-109"><a href="#cb72-109" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb72-110"><a href="#cb72-110" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-111"><a href="#cb72-111" aria-hidden="true" tabindex="-1"></a>Question: What is the p-value of the association between Yalt and X? Is it significant at the α=0.05 level?</span>
<span id="cb72-112"><a href="#cb72-112" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-113"><a href="#cb72-113" aria-hidden="true" tabindex="-1"></a><span class="fu">## Calculate the Empirical Distribution of P-values</span></span>
<span id="cb72-114"><a href="#cb72-114" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-115"><a href="#cb72-115" aria-hidden="true" tabindex="-1"></a>To understand the distribution of p-values, we need to repeat the simulation many times.</span>
<span id="cb72-116"><a href="#cb72-116" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-117"><a href="#cb72-117" aria-hidden="true" tabindex="-1"></a><span class="fu">### Convenience Function fastlm</span></span>
<span id="cb72-118"><a href="#cb72-118" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-119"><a href="#cb72-119" aria-hidden="true" tabindex="-1"></a>We'll run 10,000 regressions. A simplified function fastlm can speed this up compared to repeated calls to lm().</span>
<span id="cb72-120"><a href="#cb72-120" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-123"><a href="#cb72-123" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb72-124"><a href="#cb72-124" aria-hidden="true" tabindex="-1"></a>fastlm <span class="ot">=</span> <span class="cf">function</span>(xx,yy) {</span>
<span id="cb72-125"><a href="#cb72-125" aria-hidden="true" tabindex="-1"></a>  <span class="do">## compute betahat (regression coef) and pvalue with Ftest</span></span>
<span id="cb72-126"><a href="#cb72-126" aria-hidden="true" tabindex="-1"></a>  <span class="do">## for now it does not take covariates</span></span>
<span id="cb72-127"><a href="#cb72-127" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb72-128"><a href="#cb72-128" aria-hidden="true" tabindex="-1"></a>  df1 <span class="ot">=</span> <span class="dv">2</span> <span class="co"># DFs for model with predictor</span></span>
<span id="cb72-129"><a href="#cb72-129" aria-hidden="true" tabindex="-1"></a>  df0 <span class="ot">=</span> <span class="dv">1</span> <span class="co"># DFs for null model (intercept only)</span></span>
<span id="cb72-130"><a href="#cb72-130" aria-hidden="true" tabindex="-1"></a>  ind <span class="ot">=</span> <span class="sc">!</span><span class="fu">is.na</span>(xx) <span class="sc">&amp;</span> <span class="sc">!</span><span class="fu">is.na</span>(yy)</span>
<span id="cb72-131"><a href="#cb72-131" aria-hidden="true" tabindex="-1"></a>  xx <span class="ot">=</span> xx[ind]</span>
<span id="cb72-132"><a href="#cb72-132" aria-hidden="true" tabindex="-1"></a>  yy <span class="ot">=</span> yy[ind]</span>
<span id="cb72-133"><a href="#cb72-133" aria-hidden="true" tabindex="-1"></a>  n <span class="ot">=</span> <span class="fu">sum</span>(ind)</span>
<span id="cb72-134"><a href="#cb72-134" aria-hidden="true" tabindex="-1"></a>  xbar <span class="ot">=</span> <span class="fu">mean</span>(xx)</span>
<span id="cb72-135"><a href="#cb72-135" aria-hidden="true" tabindex="-1"></a>  ybar <span class="ot">=</span> <span class="fu">mean</span>(yy)</span>
<span id="cb72-136"><a href="#cb72-136" aria-hidden="true" tabindex="-1"></a>  xx <span class="ot">=</span> xx <span class="sc">-</span> xbar <span class="co"># Center X</span></span>
<span id="cb72-137"><a href="#cb72-137" aria-hidden="true" tabindex="-1"></a>  yy <span class="ot">=</span> yy <span class="sc">-</span> ybar <span class="co"># Center Y</span></span>
<span id="cb72-138"><a href="#cb72-138" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb72-139"><a href="#cb72-139" aria-hidden="true" tabindex="-1"></a>  SXX <span class="ot">=</span> <span class="fu">sum</span>( xx<span class="sc">^</span><span class="dv">2</span> )</span>
<span id="cb72-140"><a href="#cb72-140" aria-hidden="true" tabindex="-1"></a>  SYY <span class="ot">=</span> <span class="fu">sum</span>( yy<span class="sc">^</span><span class="dv">2</span> )</span>
<span id="cb72-141"><a href="#cb72-141" aria-hidden="true" tabindex="-1"></a>  SXY <span class="ot">=</span> <span class="fu">sum</span>( xx <span class="sc">*</span> yy )</span>
<span id="cb72-142"><a href="#cb72-142" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb72-143"><a href="#cb72-143" aria-hidden="true" tabindex="-1"></a>  betahat <span class="ot">=</span> SXY <span class="sc">/</span> SXX</span>
<span id="cb72-144"><a href="#cb72-144" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb72-145"><a href="#cb72-145" aria-hidden="true" tabindex="-1"></a>  <span class="co"># RSS1 = Residual Sum of Squares for model Y ~ X</span></span>
<span id="cb72-146"><a href="#cb72-146" aria-hidden="true" tabindex="-1"></a>  RSS1 <span class="ot">=</span> <span class="fu">sum</span>( ( yy <span class="sc">-</span> xx <span class="sc">*</span> betahat )<span class="sc">^</span><span class="dv">2</span> ) </span>
<span id="cb72-147"><a href="#cb72-147" aria-hidden="true" tabindex="-1"></a>  <span class="co"># RSS0 = Residual Sum of Squares for model Y ~ 1 (equivalent to total sum of squares of centered Y)</span></span>
<span id="cb72-148"><a href="#cb72-148" aria-hidden="true" tabindex="-1"></a>  RSS0 <span class="ot">=</span> SYY </span>
<span id="cb72-149"><a href="#cb72-149" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb72-150"><a href="#cb72-150" aria-hidden="true" tabindex="-1"></a>  <span class="co"># F-statistic comparing the two models</span></span>
<span id="cb72-151"><a href="#cb72-151" aria-hidden="true" tabindex="-1"></a>  fstat <span class="ot">=</span> ( ( RSS0 <span class="sc">-</span> RSS1 ) <span class="sc">/</span> ( df1 <span class="sc">-</span> df0 ) )  <span class="sc">/</span> ( RSS1 <span class="sc">/</span> ( n <span class="sc">-</span> df1 ) ) </span>
<span id="cb72-152"><a href="#cb72-152" aria-hidden="true" tabindex="-1"></a>  <span class="co"># P-value from F-distribution</span></span>
<span id="cb72-153"><a href="#cb72-153" aria-hidden="true" tabindex="-1"></a>  pval <span class="ot">=</span> <span class="dv">1</span> <span class="sc">-</span> <span class="fu">pf</span>(fstat, <span class="at">df1 =</span> ( df1 <span class="sc">-</span> df0 ), <span class="at">df2 =</span> ( n <span class="sc">-</span> df1 )) </span>
<span id="cb72-154"><a href="#cb72-154" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb72-155"><a href="#cb72-155" aria-hidden="true" tabindex="-1"></a>  res <span class="ot">=</span> <span class="fu">list</span>(<span class="at">betahat =</span> betahat, <span class="at">pval =</span> pval)</span>
<span id="cb72-156"><a href="#cb72-156" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(res)</span>
<span id="cb72-157"><a href="#cb72-157" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb72-158"><a href="#cb72-158" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb72-159"><a href="#cb72-159" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-160"><a href="#cb72-160" aria-hidden="true" tabindex="-1"></a><span class="fu">### Run 10,000 Simulations</span></span>
<span id="cb72-161"><a href="#cb72-161" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-162"><a href="#cb72-162" aria-hidden="true" tabindex="-1"></a>Simulate X, Ynull, and Yalt 10,000 times. Store X and the outcomes in matrices.</span>
<span id="cb72-163"><a href="#cb72-163" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-166"><a href="#cb72-166" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb72-167"><a href="#cb72-167" aria-hidden="true" tabindex="-1"></a>nsim <span class="ot">=</span> <span class="dv">10000</span> <span class="co"># Number of simulations</span></span>
<span id="cb72-168"><a href="#cb72-168" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-169"><a href="#cb72-169" aria-hidden="true" tabindex="-1"></a><span class="co"># Simulate matrices: rows=samples, cols=simulations</span></span>
<span id="cb72-170"><a href="#cb72-170" aria-hidden="true" tabindex="-1"></a>Xmat <span class="ot">=</span> <span class="fu">matrix</span>(<span class="fu">rnorm</span>(nsim <span class="sc">*</span> nsamp, <span class="at">mean=</span><span class="dv">0</span>, <span class="at">sd=</span> sigX), nsamp, nsim)</span>
<span id="cb72-171"><a href="#cb72-171" aria-hidden="true" tabindex="-1"></a>epsimat <span class="ot">=</span> <span class="fu">matrix</span>(<span class="fu">rnorm</span>(nsim <span class="sc">*</span> nsamp, <span class="at">mean=</span><span class="dv">0</span>, <span class="at">sd=</span>sigepsi), nsamp, nsim)</span>
<span id="cb72-172"><a href="#cb72-172" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-173"><a href="#cb72-173" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate Y matrices based on the formulas</span></span>
<span id="cb72-174"><a href="#cb72-174" aria-hidden="true" tabindex="-1"></a>Ymat_alt <span class="ot">=</span> Xmat <span class="sc">*</span> beta <span class="sc">+</span> epsimat </span>
<span id="cb72-175"><a href="#cb72-175" aria-hidden="true" tabindex="-1"></a>Ymat_null <span class="ot">=</span> <span class="fu">matrix</span>(<span class="fu">rnorm</span>(nsim <span class="sc">*</span> nsamp, <span class="at">mean=</span><span class="dv">0</span>, <span class="at">sd=</span>beta), nsamp, nsim) <span class="co"># Null Ys</span></span>
<span id="cb72-176"><a href="#cb72-176" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-177"><a href="#cb72-177" aria-hidden="true" tabindex="-1"></a><span class="co"># Check dimensions</span></span>
<span id="cb72-178"><a href="#cb72-178" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(Ymat_null)</span>
<span id="cb72-179"><a href="#cb72-179" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(Ymat_alt)</span>
<span id="cb72-180"><a href="#cb72-180" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-181"><a href="#cb72-181" aria-hidden="true" tabindex="-1"></a><span class="co"># Assign column names for reference</span></span>
<span id="cb72-182"><a href="#cb72-182" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(Ymat_null) <span class="ot">=</span> <span class="fu">paste0</span>(<span class="st">"sim"</span>, <span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(Ymat_null))</span>
<span id="cb72-183"><a href="#cb72-183" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(Ymat_alt) <span class="ot">=</span> <span class="fu">colnames</span>(Ymat_null)</span>
<span id="cb72-184"><a href="#cb72-184" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb72-185"><a href="#cb72-185" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-186"><a href="#cb72-186" aria-hidden="true" tabindex="-1"></a><span class="fu">### Calculate P-values Under the Null</span></span>
<span id="cb72-187"><a href="#cb72-187" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-188"><a href="#cb72-188" aria-hidden="true" tabindex="-1"></a>Run regressions for Ymat_null ~ Xmat for each simulation and store the p-values and coefficients.</span>
<span id="cb72-189"><a href="#cb72-189" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-192"><a href="#cb72-192" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb72-193"><a href="#cb72-193" aria-hidden="true" tabindex="-1"></a>pvec_null <span class="ot">=</span> <span class="fu">rep</span>(<span class="cn">NA</span>, nsim)</span>
<span id="cb72-194"><a href="#cb72-194" aria-hidden="true" tabindex="-1"></a>bvec_null <span class="ot">=</span> <span class="fu">rep</span>(<span class="cn">NA</span>, nsim)</span>
<span id="cb72-195"><a href="#cb72-195" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-196"><a href="#cb72-196" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(ss <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>nsim) {</span>
<span id="cb72-197"><a href="#cb72-197" aria-hidden="true" tabindex="-1"></a>  fit <span class="ot">=</span> <span class="fu">fastlm</span>(Xmat[,ss], Ymat_null[,ss])</span>
<span id="cb72-198"><a href="#cb72-198" aria-hidden="true" tabindex="-1"></a>  pvec_null[ss] <span class="ot">=</span> fit<span class="sc">$</span>pval  </span>
<span id="cb72-199"><a href="#cb72-199" aria-hidden="true" tabindex="-1"></a>  bvec_null[ss] <span class="ot">=</span> fit<span class="sc">$</span>betahat</span>
<span id="cb72-200"><a href="#cb72-200" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb72-201"><a href="#cb72-201" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-202"><a href="#cb72-202" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(pvec_null)</span>
<span id="cb72-203"><a href="#cb72-203" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb72-204"><a href="#cb72-204" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-207"><a href="#cb72-207" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb72-208"><a href="#cb72-208" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(pvec_null, <span class="at">xlab=</span><span class="st">"p-value"</span>, <span class="at">main=</span><span class="st">"Histogram of p-values under Null"</span>, <span class="at">breaks=</span><span class="dv">20</span>)</span>
<span id="cb72-209"><a href="#cb72-209" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb72-210"><a href="#cb72-210" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-211"><a href="#cb72-211" aria-hidden="true" tabindex="-1"></a>Questions:</span>
<span id="cb72-212"><a href="#cb72-212" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>The histogram should look approximately uniform. Why?</span>
<span id="cb72-213"><a href="#cb72-213" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>How many simulations under the null yield p-value below 0.05? What percentage is that?</span>
<span id="cb72-214"><a href="#cb72-214" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-217"><a href="#cb72-217" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb72-218"><a href="#cb72-218" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(pvec_null <span class="sc">&lt;</span> <span class="fl">0.05</span>)</span>
<span id="cb72-219"><a href="#cb72-219" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(pvec_null <span class="sc">&lt;</span> <span class="fl">0.05</span>)</span>
<span id="cb72-220"><a href="#cb72-220" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb72-221"><a href="#cb72-221" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-222"><a href="#cb72-222" aria-hidden="true" tabindex="-1"></a>What proportion of simulations do you expect to have p-values &lt; α (for any α between 0 and 1) under the null?</span>
<span id="cb72-223"><a href="#cb72-223" aria-hidden="true" tabindex="-1"></a>Why does this uniform distribution highlight the need for correction when performing many tests?</span>
<span id="cb72-224"><a href="#cb72-224" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-225"><a href="#cb72-225" aria-hidden="true" tabindex="-1"></a><span class="fu"># Bonferroni Correction</span></span>
<span id="cb72-226"><a href="#cb72-226" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-227"><a href="#cb72-227" aria-hidden="true" tabindex="-1"></a>The simplest correction is the Bonferroni correction.</span>
<span id="cb72-228"><a href="#cb72-228" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-229"><a href="#cb72-229" aria-hidden="true" tabindex="-1"></a>Method: Divide the desired significance level (e.g., 0.05) by the total number of tests performed (m, here nsim). Use this adjusted value as the new significance threshold.</span>
<span id="cb72-230"><a href="#cb72-230" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-231"><a href="#cb72-231" aria-hidden="true" tabindex="-1"></a>Goal: Controls the Family-Wise Error Rate (FWER) - the probability of calling one or more significant results across all tests under the null. P(FP&gt;=1) ≤ α.</span>
<span id="cb72-232"><a href="#cb72-232" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-235"><a href="#cb72-235" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb72-236"><a href="#cb72-236" aria-hidden="true" tabindex="-1"></a>BF_thres <span class="ot">=</span> <span class="fl">0.05</span> <span class="sc">/</span> nsim <span class="co"># Bonferroni significance threshold</span></span>
<span id="cb72-237"><a href="#cb72-237" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Bonferroni threshold:"</span>, BF_thres))</span>
<span id="cb72-238"><a href="#cb72-238" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-239"><a href="#cb72-239" aria-hidden="true" tabindex="-1"></a><span class="co"># Number of Bonferroni significant associations under the NULL</span></span>
<span id="cb72-240"><a href="#cb72-240" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Found under Null:"</span>, <span class="fu">sum</span>(pvec_null <span class="sc">&lt;</span> BF_thres)))</span>
<span id="cb72-241"><a href="#cb72-241" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Proportion under Null:"</span>, <span class="fu">mean</span>(pvec_null <span class="sc">&lt;</span> BF_thres)))</span>
<span id="cb72-242"><a href="#cb72-242" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb72-243"><a href="#cb72-243" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-244"><a href="#cb72-244" aria-hidden="true" tabindex="-1"></a>Questions:</span>
<span id="cb72-245"><a href="#cb72-245" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>What is the Bonferroni threshold for significance in this simulation?</span>
<span id="cb72-246"><a href="#cb72-246" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>How many false positives (significant results under the null) did we find using this strict threshold?</span>
<span id="cb72-247"><a href="#cb72-247" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-248"><a href="#cb72-248" aria-hidden="true" tabindex="-1"></a><span class="fu"># Mix of Null and Alternative Hypotheses</span></span>
<span id="cb72-249"><a href="#cb72-249" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-250"><a href="#cb72-250" aria-hidden="true" tabindex="-1"></a>Real-world datasets (like GWAS) contain a mixture of true associations (alternative hypothesis) and null results. Let's simulate this.</span>
<span id="cb72-251"><a href="#cb72-251" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-252"><a href="#cb72-252" aria-hidden="true" tabindex="-1"></a><span class="fu">## Create Mixed Data</span></span>
<span id="cb72-253"><a href="#cb72-253" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-254"><a href="#cb72-254" aria-hidden="true" tabindex="-1"></a>Assume a certain proportion (prop_alt) of our simulations represent true associations.</span>
<span id="cb72-255"><a href="#cb72-255" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-258"><a href="#cb72-258" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb72-259"><a href="#cb72-259" aria-hidden="true" tabindex="-1"></a>prop_alt <span class="ot">=</span> <span class="fl">0.20</span> <span class="co"># Define proportion of alternative Ys in the mixture</span></span>
<span id="cb72-260"><a href="#cb72-260" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-261"><a href="#cb72-261" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a selection vector: 1 if alternative, 0 if null</span></span>
<span id="cb72-262"><a href="#cb72-262" aria-hidden="true" tabindex="-1"></a>selectvec <span class="ot">=</span> <span class="fu">rbinom</span>(nsim, <span class="dv">1</span>, prop_alt) </span>
<span id="cb72-263"><a href="#cb72-263" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(selectvec) <span class="ot">=</span> <span class="fu">colnames</span>(Ymat_alt) <span class="co"># Keep track</span></span>
<span id="cb72-264"><a href="#cb72-264" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(selectvec) <span class="co"># Show counts of null (0) vs alternative (1)</span></span>
<span id="cb72-265"><a href="#cb72-265" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-266"><a href="#cb72-266" aria-hidden="true" tabindex="-1"></a><span class="co"># Create the mixed Y matrix</span></span>
<span id="cb72-267"><a href="#cb72-267" aria-hidden="true" tabindex="-1"></a><span class="co"># If selectvec[i] is 1, use Ymat_alt[,i]; if 0, use Ymat_null[,i]</span></span>
<span id="cb72-268"><a href="#cb72-268" aria-hidden="true" tabindex="-1"></a>Ymat_mix <span class="ot">=</span> <span class="fu">sweep</span>(Ymat_alt, <span class="dv">2</span>, selectvec, <span class="at">FUN=</span><span class="st">'*'</span>) <span class="sc">+</span> <span class="fu">sweep</span>(Ymat_null, <span class="dv">2</span>, <span class="dv">1</span> <span class="sc">-</span> selectvec, <span class="at">FUN=</span><span class="st">'*'</span>)</span>
<span id="cb72-269"><a href="#cb72-269" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb72-270"><a href="#cb72-270" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-271"><a href="#cb72-271" aria-hidden="true" tabindex="-1"></a><span class="fu">## Calculate P-values for Mixed Data</span></span>
<span id="cb72-272"><a href="#cb72-272" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-273"><a href="#cb72-273" aria-hidden="true" tabindex="-1"></a>Run regressions for Ymat_mix ~ Xmat.</span>
<span id="cb72-274"><a href="#cb72-274" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-277"><a href="#cb72-277" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb72-278"><a href="#cb72-278" aria-hidden="true" tabindex="-1"></a>pvec_mix <span class="ot">=</span> <span class="fu">rep</span>(<span class="cn">NA</span>, nsim)</span>
<span id="cb72-279"><a href="#cb72-279" aria-hidden="true" tabindex="-1"></a>bvec_mix <span class="ot">=</span> <span class="fu">rep</span>(<span class="cn">NA</span>, nsim)</span>
<span id="cb72-280"><a href="#cb72-280" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-281"><a href="#cb72-281" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(ss <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>nsim) {</span>
<span id="cb72-282"><a href="#cb72-282" aria-hidden="true" tabindex="-1"></a>  fit <span class="ot">=</span> <span class="fu">fastlm</span>(Xmat[,ss], Ymat_mix[,ss])</span>
<span id="cb72-283"><a href="#cb72-283" aria-hidden="true" tabindex="-1"></a>  pvec_mix[ss] <span class="ot">=</span> fit<span class="sc">$</span>pval  </span>
<span id="cb72-284"><a href="#cb72-284" aria-hidden="true" tabindex="-1"></a>  bvec_mix[ss] <span class="ot">=</span> fit<span class="sc">$</span>betahat</span>
<span id="cb72-285"><a href="#cb72-285" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb72-286"><a href="#cb72-286" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-287"><a href="#cb72-287" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(pvec_mix)</span>
<span id="cb72-288"><a href="#cb72-288" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb72-289"><a href="#cb72-289" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-292"><a href="#cb72-292" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb72-293"><a href="#cb72-293" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(pvec_mix, <span class="at">xlab=</span><span class="st">"p-value"</span>, <span class="at">main=</span><span class="st">"Histogram of p-values under mixture"</span>, <span class="at">breaks=</span><span class="dv">20</span>)</span>
<span id="cb72-294"><a href="#cb72-294" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb72-295"><a href="#cb72-295" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-296"><a href="#cb72-296" aria-hidden="true" tabindex="-1"></a><span class="fu">## Interpreting Mixed P-values</span></span>
<span id="cb72-297"><a href="#cb72-297" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-298"><a href="#cb72-298" aria-hidden="true" tabindex="-1"></a>The histogram shows a spike near zero (from true alternatives) superimposed on a uniform background (from true nulls).</span>
<span id="cb72-299"><a href="#cb72-299" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-302"><a href="#cb72-302" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb72-303"><a href="#cb72-303" aria-hidden="true" tabindex="-1"></a>thres <span class="ot">=</span> <span class="fl">0.05</span></span>
<span id="cb72-304"><a href="#cb72-304" aria-hidden="true" tabindex="-1"></a>m_signif <span class="ot">=</span> <span class="fu">sum</span>(pvec_mix <span class="sc">&lt;</span> thres)      <span class="co"># Observed number of significant associations</span></span>
<span id="cb72-305"><a href="#cb72-305" aria-hidden="true" tabindex="-1"></a>m_expected_null <span class="ot">=</span> thres <span class="sc">*</span> <span class="fu">sum</span>(selectvec <span class="sc">==</span> <span class="dv">0</span>) <span class="co"># Expected number of FPs from nulls</span></span>
<span id="cb72-306"><a href="#cb72-306" aria-hidden="true" tabindex="-1"></a>m_expected_all_null <span class="ot">=</span> thres <span class="sc">*</span> nsim <span class="co"># Expected if *all* were null</span></span>
<span id="cb72-307"><a href="#cb72-307" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-308"><a href="#cb72-308" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Observed significant (p&lt;0.05):"</span>, m_signif))</span>
<span id="cb72-309"><a href="#cb72-309" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Expected false positives if all were null:"</span>, m_expected_all_null))</span>
<span id="cb72-310"><a href="#cb72-310" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Expected false positives from the actual nulls:"</span>, <span class="fu">round</span>(m_expected_null)))</span>
<span id="cb72-311"><a href="#cb72-311" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb72-312"><a href="#cb72-312" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-313"><a href="#cb72-313" aria-hidden="true" tabindex="-1"></a>We observed many more significant results than expected if all tests were null. How can we estimate the proportion of false discoveries among those we call significant?</span>
<span id="cb72-314"><a href="#cb72-314" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-315"><a href="#cb72-315" aria-hidden="true" tabindex="-1"></a><span class="fu">## False Discovery Rate (FDR) Estimate (using known truth)</span></span>
<span id="cb72-316"><a href="#cb72-316" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-317"><a href="#cb72-317" aria-hidden="true" tabindex="-1"></a>Since this is a simulation, we know which tests were truly null (selectvec == 0). We can calculate the actual FDR for a given p-value threshold.</span>
<span id="cb72-318"><a href="#cb72-318" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-321"><a href="#cb72-321" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb72-322"><a href="#cb72-322" aria-hidden="true" tabindex="-1"></a><span class="co"># FDR = (Number of significant tests that are truly null) / (Total number of significant tests)</span></span>
<span id="cb72-323"><a href="#cb72-323" aria-hidden="true" tabindex="-1"></a>FP <span class="ot">=</span> <span class="fu">sum</span>(pvec_mix <span class="sc">&lt;</span> thres <span class="sc">&amp;</span> selectvec <span class="sc">==</span> <span class="dv">0</span>) <span class="co"># False Positives</span></span>
<span id="cb72-324"><a href="#cb72-324" aria-hidden="true" tabindex="-1"></a>TP <span class="ot">=</span> <span class="fu">sum</span>(pvec_mix <span class="sc">&lt;</span> thres <span class="sc">&amp;</span> selectvec <span class="sc">==</span> <span class="dv">1</span>) <span class="co"># True Positives</span></span>
<span id="cb72-325"><a href="#cb72-325" aria-hidden="true" tabindex="-1"></a>S <span class="ot">=</span> <span class="fu">sum</span>(pvec_mix <span class="sc">&lt;</span> thres) <span class="co"># Total Significant (S = FP + TP)</span></span>
<span id="cb72-326"><a href="#cb72-326" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-327"><a href="#cb72-327" aria-hidden="true" tabindex="-1"></a>FDR_sim <span class="ot">=</span> FP <span class="sc">/</span> S </span>
<span id="cb72-328"><a href="#cb72-328" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Simulated FDR at p &lt;"</span>, thres, <span class="st">":"</span>, <span class="fu">round</span>(FDR_sim, <span class="dv">4</span>)))</span>
<span id="cb72-329"><a href="#cb72-329" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb72-330"><a href="#cb72-330" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-331"><a href="#cb72-331" aria-hidden="true" tabindex="-1"></a>This means that if we used p &lt; 0.05 as our criterion, about round(FDR_sim*100, 1)% of our significant findings would actually be false positives.</span>
<span id="cb72-332"><a href="#cb72-332" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-333"><a href="#cb72-333" aria-hidden="true" tabindex="-1"></a>Questions:</span>
<span id="cb72-334"><a href="#cb72-334" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>What is the proportion of false discoveries (FDR) if we use a significance level of 0.01?</span>
<span id="cb72-335"><a href="#cb72-335" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-338"><a href="#cb72-338" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb72-339"><a href="#cb72-339" aria-hidden="true" tabindex="-1"></a>thres_01 <span class="ot">=</span> <span class="fl">0.01</span></span>
<span id="cb72-340"><a href="#cb72-340" aria-hidden="true" tabindex="-1"></a>FP_01 <span class="ot">=</span> <span class="fu">sum</span>(pvec_mix <span class="sc">&lt;</span> thres_01 <span class="sc">&amp;</span> selectvec <span class="sc">==</span> <span class="dv">0</span>)</span>
<span id="cb72-341"><a href="#cb72-341" aria-hidden="true" tabindex="-1"></a>S_01 <span class="ot">=</span> <span class="fu">sum</span>(pvec_mix <span class="sc">&lt;</span> thres_01)</span>
<span id="cb72-342"><a href="#cb72-342" aria-hidden="true" tabindex="-1"></a>FDR_sim_01 <span class="ot">=</span> FP_01 <span class="sc">/</span> S_01</span>
<span id="cb72-343"><a href="#cb72-343" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Simulated FDR at p &lt; 0.01:"</span>, <span class="fu">round</span>(FDR_sim_01, <span class="dv">4</span>)))</span>
<span id="cb72-344"><a href="#cb72-344" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb72-345"><a href="#cb72-345" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-346"><a href="#cb72-346" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>What is the proportion of false discoveries if we use the Bonferroni threshold?</span>
<span id="cb72-347"><a href="#cb72-347" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-350"><a href="#cb72-350" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb72-351"><a href="#cb72-351" aria-hidden="true" tabindex="-1"></a>FP_bf <span class="ot">=</span> <span class="fu">sum</span>(pvec_mix <span class="sc">&lt;</span> BF_thres <span class="sc">&amp;</span> selectvec <span class="sc">==</span> <span class="dv">0</span>)</span>
<span id="cb72-352"><a href="#cb72-352" aria-hidden="true" tabindex="-1"></a>S_bf <span class="ot">=</span> <span class="fu">sum</span>(pvec_mix <span class="sc">&lt;</span> BF_thres)</span>
<span id="cb72-353"><a href="#cb72-353" aria-hidden="true" tabindex="-1"></a>FDR_sim_bf <span class="ot">=</span> <span class="fu">ifelse</span>(S_bf <span class="sc">&gt;</span> <span class="dv">0</span>, FP_bf <span class="sc">/</span> S_bf, <span class="dv">0</span>) <span class="co"># Avoid division by zero</span></span>
<span id="cb72-354"><a href="#cb72-354" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Simulated FDR at Bonferroni p &lt;"</span>, BF_thres, <span class="st">":"</span>, <span class="fu">round</span>(FDR_sim_bf, <span class="dv">4</span>)))</span>
<span id="cb72-355"><a href="#cb72-355" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Number significant at Bonferroni:"</span>, S_bf))</span>
<span id="cb72-356"><a href="#cb72-356" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb72-357"><a href="#cb72-357" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-358"><a href="#cb72-358" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>What is the proportion of missed signals (False Negative Rate among true alternatives) using the Bonferroni threshold?</span>
<span id="cb72-359"><a href="#cb72-359" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-362"><a href="#cb72-362" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb72-363"><a href="#cb72-363" aria-hidden="true" tabindex="-1"></a>FN_bf <span class="ot">=</span> <span class="fu">sum</span>(pvec_mix <span class="sc">&gt;=</span> BF_thres <span class="sc">&amp;</span> selectvec <span class="sc">==</span> <span class="dv">1</span>) <span class="co"># False Negatives</span></span>
<span id="cb72-364"><a href="#cb72-364" aria-hidden="true" tabindex="-1"></a>Total_Alt <span class="ot">=</span> <span class="fu">sum</span>(selectvec <span class="sc">==</span> <span class="dv">1</span>) <span class="co"># Total true alternatives</span></span>
<span id="cb72-365"><a href="#cb72-365" aria-hidden="true" tabindex="-1"></a>FNR_sim_bf <span class="ot">=</span> FN_bf <span class="sc">/</span> Total_Alt</span>
<span id="cb72-366"><a href="#cb72-366" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Simulated FNR at Bonferroni p &lt;"</span>, BF_thres, <span class="st">":"</span>, <span class="fu">round</span>(FNR_sim_bf, <span class="dv">4</span>)))</span>
<span id="cb72-367"><a href="#cb72-367" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb72-368"><a href="#cb72-368" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-369"><a href="#cb72-369" aria-hidden="true" tabindex="-1"></a><span class="fu"># Common Approaches to Control Type I Errors</span></span>
<span id="cb72-370"><a href="#cb72-370" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-371"><a href="#cb72-371" aria-hidden="true" tabindex="-1"></a>Let m be the total number of tests. Consider the following table:</span>
<span id="cb72-372"><a href="#cb72-372" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-373"><a href="#cb72-373" aria-hidden="true" tabindex="-1"></a>|| Called Not Significant | Called Significant | Total</span>
<span id="cb72-374"><a href="#cb72-374" aria-hidden="true" tabindex="-1"></a>|---|---------------|---------------------|-----</span>
<span id="cb72-375"><a href="#cb72-375" aria-hidden="true" tabindex="-1"></a>| Null True | TN | FP | $m_0$</span>
<span id="cb72-376"><a href="#cb72-376" aria-hidden="true" tabindex="-1"></a>| Alternative True | FN | TP | $m_1$</span>
<span id="cb72-377"><a href="#cb72-377" aria-hidden="true" tabindex="-1"></a>| Total | $m$ - S | S | $m$</span>
<span id="cb72-378"><a href="#cb72-378" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-379"><a href="#cb72-379" aria-hidden="true" tabindex="-1"></a>(m total # tests, $m_0$ # null tests, FP false positives, TP true positives, TN true negatives, FN false negatives, S total significant)</span>
<span id="cb72-380"><a href="#cb72-380" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-381"><a href="#cb72-381" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>FWER (Family-Wise Error Rate): P(FP ≥ 1). Probability of at least one false positive. Controlled by Bonferroni. Often too strict.</span>
<span id="cb72-382"><a href="#cb72-382" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>FDR (False Discovery Rate): E<span class="co">[</span><span class="ot">FP / S | S &gt; 0</span><span class="co">]</span>. Expected proportion of false positives among all significant results. A more liberal approach, often preferred when many tests are performed and finding true positives is important.</span>
<span id="cb72-383"><a href="#cb72-383" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-384"><a href="#cb72-384" aria-hidden="true" tabindex="-1"></a><span class="fu">## Table of Simulation Results</span></span>
<span id="cb72-385"><a href="#cb72-385" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-386"><a href="#cb72-386" aria-hidden="true" tabindex="-1"></a>Let's construct the confusion table for our mixed simulation using alpha = 0.05.</span>
<span id="cb72-387"><a href="#cb72-387" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-390"><a href="#cb72-390" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb72-391"><a href="#cb72-391" aria-hidden="true" tabindex="-1"></a>count_table <span class="ot">=</span> <span class="fu">t</span>(<span class="fu">table</span>(<span class="at">Significant =</span> pvec_mix <span class="sc">&lt;</span> <span class="fl">0.05</span>, <span class="at">True_Status =</span> selectvec))</span>
<span id="cb72-392"><a href="#cb72-392" aria-hidden="true" tabindex="-1"></a><span class="co"># Reorder and rename for standard confusion matrix format</span></span>
<span id="cb72-393"><a href="#cb72-393" aria-hidden="true" tabindex="-1"></a><span class="co"># Rows: True Status (0=Null, 1=Alt); Columns: Called Status (FALSE=Not Sig, TRUE=Sig)</span></span>
<span id="cb72-394"><a href="#cb72-394" aria-hidden="true" tabindex="-1"></a>conf_matrix <span class="ot">=</span> count_table[, <span class="fu">c</span>(<span class="st">"FALSE"</span>, <span class="st">"TRUE"</span>)] </span>
<span id="cb72-395"><a href="#cb72-395" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(conf_matrix) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"Called Not Significant"</span>, <span class="st">"Called Significant"</span>)</span>
<span id="cb72-396"><a href="#cb72-396" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(conf_matrix) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"Null True"</span>, <span class="st">"Alt True"</span>)</span>
<span id="cb72-397"><a href="#cb72-397" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-398"><a href="#cb72-398" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(conf_matrix, <span class="at">caption=</span><span class="st">"Confusion Matrix at p &lt; 0.05"</span>)</span>
<span id="cb72-399"><a href="#cb72-399" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-400"><a href="#cb72-400" aria-hidden="true" tabindex="-1"></a><span class="co"># Extracting values</span></span>
<span id="cb72-401"><a href="#cb72-401" aria-hidden="true" tabindex="-1"></a>TN <span class="ot">=</span> conf_matrix[<span class="st">"Null True"</span>, <span class="st">"Called Not Significant"</span>]</span>
<span id="cb72-402"><a href="#cb72-402" aria-hidden="true" tabindex="-1"></a>FP <span class="ot">=</span> conf_matrix[<span class="st">"Null True"</span>, <span class="st">"Called Significant"</span>]</span>
<span id="cb72-403"><a href="#cb72-403" aria-hidden="true" tabindex="-1"></a>FN <span class="ot">=</span> conf_matrix[<span class="st">"Alt True"</span>, <span class="st">"Called Not Significant"</span>]</span>
<span id="cb72-404"><a href="#cb72-404" aria-hidden="true" tabindex="-1"></a>TP <span class="ot">=</span> conf_matrix[<span class="st">"Alt True"</span>, <span class="st">"Called Significant"</span>]</span>
<span id="cb72-405"><a href="#cb72-405" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"TN:"</span>, TN, <span class="st">" FP:"</span>, FP, <span class="st">" FN:"</span>, FN, <span class="st">" TP:"</span>, TP))</span>
<span id="cb72-406"><a href="#cb72-406" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb72-407"><a href="#cb72-407" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-408"><a href="#cb72-408" aria-hidden="true" tabindex="-1"></a><span class="fu"># Use qvalue Package to Estimate FDR</span></span>
<span id="cb72-409"><a href="#cb72-409" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-410"><a href="#cb72-410" aria-hidden="true" tabindex="-1"></a>In real analysis, we don't know selectvec. The qvalue package estimates FDR and related quantities directly from the p-value distribution.</span>
<span id="cb72-411"><a href="#cb72-411" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-412"><a href="#cb72-412" aria-hidden="true" tabindex="-1"></a>q-value: The minimum FDR at which a test would be called significant. If you call all tests with q ≤ x significant, you expect the FDR to be approximately x.</span>
<span id="cb72-413"><a href="#cb72-413" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-416"><a href="#cb72-416" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb72-417"><a href="#cb72-417" aria-hidden="true" tabindex="-1"></a><span class="co"># Install qvalue package if you haven't already</span></span>
<span id="cb72-418"><a href="#cb72-418" aria-hidden="true" tabindex="-1"></a><span class="co"># if (!requireNamespace("BiocManager", quietly = TRUE))</span></span>
<span id="cb72-419"><a href="#cb72-419" aria-hidden="true" tabindex="-1"></a><span class="co">#    install.packages("BiocManager")</span></span>
<span id="cb72-420"><a href="#cb72-420" aria-hidden="true" tabindex="-1"></a><span class="co">#BiocManager::install("qvalue")</span></span>
<span id="cb72-421"><a href="#cb72-421" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb72-422"><a href="#cb72-422" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-425"><a href="#cb72-425" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb72-426"><a href="#cb72-426" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(qvalue)</span>
<span id="cb72-427"><a href="#cb72-427" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb72-428"><a href="#cb72-428" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-431"><a href="#cb72-431" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb72-432"><a href="#cb72-432" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate q-values for the mixed p-values</span></span>
<span id="cb72-433"><a href="#cb72-433" aria-hidden="true" tabindex="-1"></a>qres_mix <span class="ot">=</span> <span class="fu">qvalue</span>(pvec_mix)</span>
<span id="cb72-434"><a href="#cb72-434" aria-hidden="true" tabindex="-1"></a>qvec_mix <span class="ot">=</span> qres_mix<span class="sc">$</span>qvalues</span>
<span id="cb72-435"><a href="#cb72-435" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-436"><a href="#cb72-436" aria-hidden="true" tabindex="-1"></a><span class="co"># Also calculate for the null p-values for comparison</span></span>
<span id="cb72-437"><a href="#cb72-437" aria-hidden="true" tabindex="-1"></a>qres_null <span class="ot">=</span> <span class="fu">qvalue</span>(pvec_null)</span>
<span id="cb72-438"><a href="#cb72-438" aria-hidden="true" tabindex="-1"></a>qvec_null <span class="ot">=</span> qres_null<span class="sc">$</span>qvalues</span>
<span id="cb72-439"><a href="#cb72-439" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-440"><a href="#cb72-440" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(qvec_mix)</span>
<span id="cb72-441"><a href="#cb72-441" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb72-442"><a href="#cb72-442" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-443"><a href="#cb72-443" aria-hidden="true" tabindex="-1"></a><span class="fu">## Visualize Q-values</span></span>
<span id="cb72-444"><a href="#cb72-444" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-445"><a href="#cb72-445" aria-hidden="true" tabindex="-1"></a>Let's see if small q-values correspond to true associations (selectvec = 1).</span>
<span id="cb72-446"><a href="#cb72-446" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-449"><a href="#cb72-449" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb72-450"><a href="#cb72-450" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(qvec_mix <span class="sc">~</span> selectvec, <span class="at">main=</span><span class="st">'Q-value by True Status'</span>, <span class="at">xlab=</span><span class="st">"True Status (0=Null, 1=Alt)"</span>, <span class="at">ylab=</span><span class="st">"Q-value"</span>)</span>
<span id="cb72-451"><a href="#cb72-451" aria-hidden="true" tabindex="-1"></a><span class="fu">grid</span>()</span>
<span id="cb72-452"><a href="#cb72-452" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb72-453"><a href="#cb72-453" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-454"><a href="#cb72-454" aria-hidden="true" tabindex="-1"></a>Plot sorted q-values, colored by true status.</span>
<span id="cb72-455"><a href="#cb72-455" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-458"><a href="#cb72-458" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb72-459"><a href="#cb72-459" aria-hidden="true" tabindex="-1"></a>ind <span class="ot">=</span> <span class="fu">order</span>(qvec_mix, <span class="at">decreasing=</span><span class="cn">FALSE</span>)</span>
<span id="cb72-460"><a href="#cb72-460" aria-hidden="true" tabindex="-1"></a><span class="co"># Using selectvec*2 + 1 maps 0-&gt;1 (black) and 1-&gt;3 (green) for plotting colors</span></span>
<span id="cb72-461"><a href="#cb72-461" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">sort</span>(qvec_mix), <span class="at">col=</span>selectvec[ind]<span class="sc">*</span><span class="dv">2</span> <span class="sc">+</span> <span class="dv">1</span>, </span>
<span id="cb72-462"><a href="#cb72-462" aria-hidden="true" tabindex="-1"></a>     <span class="at">pch=</span><span class="dv">16</span>, <span class="co">#selectvec[ind] + 1, </span></span>
<span id="cb72-463"><a href="#cb72-463" aria-hidden="true" tabindex="-1"></a>     <span class="co"># lwd=selectvec[ind]*2 + 1,</span></span>
<span id="cb72-464"><a href="#cb72-464" aria-hidden="true" tabindex="-1"></a>     <span class="at">main=</span><span class="st">"Sorted Q-values by True Status"</span>,</span>
<span id="cb72-465"><a href="#cb72-465" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab=</span><span class="st">"Tests ordered by Q-value"</span>, <span class="at">ylab=</span><span class="st">"Q-value"</span>)</span>
<span id="cb72-466"><a href="#cb72-466" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"bottomright"</span>, <span class="at">legend=</span><span class="fu">c</span>(<span class="st">"True Null"</span>, <span class="st">"True Alternative"</span>), <span class="at">pch=</span><span class="dv">16</span>, <span class="at">col=</span><span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>))</span>
<span id="cb72-467"><a href="#cb72-467" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb72-468"><a href="#cb72-468" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-469"><a href="#cb72-469" aria-hidden="true" tabindex="-1"></a>Questions:</span>
<span id="cb72-470"><a href="#cb72-470" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Do small q-values tend to correspond to true alternatives (green points)?</span>
<span id="cb72-471"><a href="#cb72-471" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Interpret the sorted q-value plot: What does the pattern show?</span>
<span id="cb72-472"><a href="#cb72-472" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-473"><a href="#cb72-473" aria-hidden="true" tabindex="-1"></a>Compare p-value and q-value distributions by true status.</span>
<span id="cb72-474"><a href="#cb72-474" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-477"><a href="#cb72-477" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb72-478"><a href="#cb72-478" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb72-479"><a href="#cb72-479" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(pvec_mix <span class="sc">~</span> selectvec, <span class="at">main=</span><span class="st">'P-value by True Status'</span>, <span class="at">xlab=</span><span class="st">"True Status (0=Null, 1=Alt)"</span>, <span class="at">ylab=</span><span class="st">"P-value"</span>); <span class="fu">grid</span>()</span>
<span id="cb72-480"><a href="#cb72-480" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(qvec_mix <span class="sc">~</span> selectvec, <span class="at">main=</span><span class="st">'Q-value by True Status'</span>, <span class="at">xlab=</span><span class="st">"True Status (0=Null, 1=Alt)"</span>, <span class="at">ylab=</span><span class="st">"Q-value"</span>); <span class="fu">grid</span>()</span>
<span id="cb72-481"><a href="#cb72-481" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span>
<span id="cb72-482"><a href="#cb72-482" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb72-483"><a href="#cb72-483" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-484"><a href="#cb72-484" aria-hidden="true" tabindex="-1"></a>Question: Interpret these boxplots. How do q-values differ from p-values, especially for the null tests?</span>
<span id="cb72-485"><a href="#cb72-485" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-486"><a href="#cb72-486" aria-hidden="true" tabindex="-1"></a><span class="fu">## How Do Q-values and P-values Relate?</span></span>
<span id="cb72-487"><a href="#cb72-487" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-490"><a href="#cb72-490" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb72-491"><a href="#cb72-491" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))</span>
<span id="cb72-492"><a href="#cb72-492" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pvec_null, qvec_null, <span class="at">main=</span><span class="st">'Q-value vs P-value (All Null)'</span>, <span class="at">xlab=</span><span class="st">"P-value"</span>, <span class="at">ylab=</span><span class="st">"Q-value"</span>); <span class="fu">grid</span>(); <span class="fu">abline</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="at">col=</span><span class="st">'red'</span>,<span class="at">lty=</span><span class="dv">2</span>)</span>
<span id="cb72-493"><a href="#cb72-493" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pvec_mix, qvec_mix, <span class="at">main=</span><span class="st">'Q-value vs P-value (Mixture)'</span>, <span class="at">xlab=</span><span class="st">"P-value"</span>, <span class="at">ylab=</span><span class="st">"Q-value"</span>); <span class="fu">grid</span>(); <span class="fu">abline</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="at">col=</span><span class="st">'red'</span>,<span class="at">lty=</span><span class="dv">2</span>)</span>
<span id="cb72-494"><a href="#cb72-494" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))</span>
<span id="cb72-495"><a href="#cb72-495" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb72-496"><a href="#cb72-496" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-497"><a href="#cb72-497" aria-hidden="true" tabindex="-1"></a>Q-values are monotonically increasing functions of p-values. Note that q-values are often ≥ p-values. The q-value can be interpreted as the FDR associated with calling that specific test significant.</span>
<span id="cb72-498"><a href="#cb72-498" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-499"><a href="#cb72-499" aria-hidden="true" tabindex="-1"></a>Questions:</span>
<span id="cb72-500"><a href="#cb72-500" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>What is the smallest q-value when all simulations are from the null? Why?</span>
<span id="cb72-501"><a href="#cb72-501" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>What is the smallest q-value when simulations are from the mixture?</span>
<span id="cb72-502"><a href="#cb72-502" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-503"><a href="#cb72-503" aria-hidden="true" tabindex="-1"></a><span class="fu">## Estimating π₀ (Proportion of True Nulls)</span></span>
<span id="cb72-504"><a href="#cb72-504" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-505"><a href="#cb72-505" aria-hidden="true" tabindex="-1"></a>The qvalue package also estimates π₀, the overall proportion of features (tests) that are truly null. </span>
<span id="cb72-506"><a href="#cb72-506" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-507"><a href="#cb72-507" aria-hidden="true" tabindex="-1"></a>π₁ = 1 - π₀ is the proportion of true alternative hypotheses.</span>
<span id="cb72-508"><a href="#cb72-508" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-511"><a href="#cb72-511" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb72-512"><a href="#cb72-512" aria-hidden="true" tabindex="-1"></a><span class="co"># Estimated pi0 when all tests were truly null</span></span>
<span id="cb72-513"><a href="#cb72-513" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Estimated pi0 (all null):"</span>, <span class="fu">round</span>(qres_null<span class="sc">$</span>pi0, <span class="dv">4</span>)))</span>
<span id="cb72-514"><a href="#cb72-514" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-515"><a href="#cb72-515" aria-hidden="true" tabindex="-1"></a><span class="co"># Estimated pi0 for the mixed data</span></span>
<span id="cb72-516"><a href="#cb72-516" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"Estimated pi0 (mixture):"</span>, <span class="fu">round</span>(qres_mix<span class="sc">$</span>pi0, <span class="dv">4</span>)))</span>
<span id="cb72-517"><a href="#cb72-517" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-518"><a href="#cb72-518" aria-hidden="true" tabindex="-1"></a><span class="co"># True proportion used in simulation</span></span>
<span id="cb72-519"><a href="#cb72-519" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">"True pi0 used:"</span>, <span class="dv">1</span> <span class="sc">-</span> prop_alt))</span>
<span id="cb72-520"><a href="#cb72-520" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb72-521"><a href="#cb72-521" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-522"><a href="#cb72-522" aria-hidden="true" tabindex="-1"></a>Question: How well did the qvalue package estimate the true proportion of null hypotheses (pi0) in our mixed simulation?</span>
<span id="cb72-523"><a href="#cb72-523" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-524"><a href="#cb72-524" aria-hidden="true" tabindex="-1"></a><span class="fu"># References</span></span>
<span id="cb72-525"><a href="#cb72-525" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-526"><a href="#cb72-526" aria-hidden="true" tabindex="-1"></a>Storey, John D., and Robert Tibshirani. 2003. "Statistical Significance for Genomewide Studies." Proceedings of the National Academy of Sciences 100 (16): 9440–45.</span>
</code><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->




</body></html>